{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "OryzaGP.ipynb",
      "provenance": [],
      "authorship_tag": "ABX9TyPUC32XC9O9qDvoIpoXkcw4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pierrelarmande/OryzaGP/blob/main/src/OryzaGP.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "HhCo9j5-Y3vN"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import json\n",
        "import csv\n",
        "import pickle\n",
        "from urllib.request import urlopen\n",
        "import numpy as np\n",
        "import missingno as msno\n",
        "import nltk\n",
        "import pprint\n",
        "import os, sys, re, pprint, datetime, time\n",
        "import copy"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install scispacy\n",
        "!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.0/en_core_sci_sm-0.5.0.tar.gz"
      ],
      "metadata": {
        "id": "Pujom8sCsj0E",
        "outputId": "a52860a2-2073-4e6a-e01c-922da09aeada",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting scispacy\n",
            "  Downloading scispacy-0.5.0-py3-none-any.whl (44 kB)\n",
            "\u001b[K     |████████████████████████████████| 44 kB 2.1 MB/s \n",
            "\u001b[?25hRequirement already satisfied: scikit-learn>=0.20.3 in /usr/local/lib/python3.7/dist-packages (from scispacy) (1.0.2)\n",
            "Collecting pysbd\n",
            "  Downloading pysbd-0.3.4-py3-none-any.whl (71 kB)\n",
            "\u001b[K     |████████████████████████████████| 71 kB 4.8 MB/s \n",
            "\u001b[?25hRequirement already satisfied: joblib in /usr/local/lib/python3.7/dist-packages (from scispacy) (1.1.0)\n",
            "Collecting nmslib>=1.7.3.6\n",
            "  Downloading nmslib-2.1.1-cp37-cp37m-manylinux2010_x86_64.whl (13.5 MB)\n",
            "\u001b[K     |████████████████████████████████| 13.5 MB 27.2 MB/s \n",
            "\u001b[?25hCollecting spacy<3.3.0,>=3.2.0\n",
            "  Downloading spacy-3.2.4-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.0 MB)\n",
            "\u001b[K     |████████████████████████████████| 6.0 MB 52.7 MB/s \n",
            "\u001b[?25hCollecting conllu\n",
            "  Downloading conllu-4.5.2-py2.py3-none-any.whl (16 kB)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scispacy) (2.23.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from scispacy) (1.21.6)\n",
            "Collecting pybind11<2.6.2\n",
            "  Downloading pybind11-2.6.1-py2.py3-none-any.whl (188 kB)\n",
            "\u001b[K     |████████████████████████████████| 188 kB 68.0 MB/s \n",
            "\u001b[?25hRequirement already satisfied: psutil in /usr/local/lib/python3.7/dist-packages (from nmslib>=1.7.3.6->scispacy) (5.4.8)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (1.24.3)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (2.10)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (3.0.4)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (2022.6.15)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.20.3->scispacy) (3.1.0)\n",
            "Requirement already satisfied: scipy>=1.1.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.20.3->scispacy) (1.7.3)\n",
            "Collecting thinc<8.1.0,>=8.0.12\n",
            "  Downloading thinc-8.0.17-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (660 kB)\n",
            "\u001b[K     |████████████████████████████████| 660 kB 53.8 MB/s \n",
            "\u001b[?25hCollecting pydantic!=1.8,!=1.8.1,<1.9.0,>=1.7.4\n",
            "  Downloading pydantic-1.8.2-cp37-cp37m-manylinux2014_x86_64.whl (10.1 MB)\n",
            "\u001b[K     |████████████████████████████████| 10.1 MB 48.4 MB/s \n",
            "\u001b[?25hRequirement already satisfied: langcodes<4.0.0,>=3.2.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (3.3.0)\n",
            "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.8 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (3.0.10)\n",
            "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (2.0.8)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (2.11.3)\n",
            "Requirement already satisfied: pathy>=0.3.5 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (0.6.2)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (57.4.0)\n",
            "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (1.0.3)\n",
            "Requirement already satisfied: srsly<3.0.0,>=2.4.1 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (2.4.4)\n",
            "Collecting typing-extensions<4.0.0.0,>=3.7.4\n",
            "  Downloading typing_extensions-3.10.0.2-py3-none-any.whl (26 kB)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (1.0.8)\n",
            "Requirement already satisfied: typer<0.5.0,>=0.3.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (0.4.2)\n",
            "Requirement already satisfied: wasabi<1.1.0,>=0.8.1 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (0.10.1)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (2.0.6)\n",
            "Requirement already satisfied: click<8.1.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (7.1.2)\n",
            "Requirement already satisfied: blis<0.8.0,>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (0.7.8)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (21.3)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (3.0.7)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (4.64.0)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from catalogue<2.1.0,>=2.0.6->spacy<3.3.0,>=3.2.0->scispacy) (3.8.1)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging>=20.0->spacy<3.3.0,>=3.2.0->scispacy) (3.0.9)\n",
            "Requirement already satisfied: smart-open<6.0.0,>=5.2.1 in /usr/local/lib/python3.7/dist-packages (from pathy>=0.3.5->spacy<3.3.0,>=3.2.0->scispacy) (5.2.1)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.7/dist-packages (from jinja2->spacy<3.3.0,>=3.2.0->scispacy) (2.0.1)\n",
            "Installing collected packages: typing-extensions, pydantic, thinc, pybind11, spacy, pysbd, nmslib, conllu, scispacy\n",
            "  Attempting uninstall: typing-extensions\n",
            "    Found existing installation: typing-extensions 4.1.1\n",
            "    Uninstalling typing-extensions-4.1.1:\n",
            "      Successfully uninstalled typing-extensions-4.1.1\n",
            "  Attempting uninstall: pydantic\n",
            "    Found existing installation: pydantic 1.9.2\n",
            "    Uninstalling pydantic-1.9.2:\n",
            "      Successfully uninstalled pydantic-1.9.2\n",
            "  Attempting uninstall: thinc\n",
            "    Found existing installation: thinc 8.1.0\n",
            "    Uninstalling thinc-8.1.0:\n",
            "      Successfully uninstalled thinc-8.1.0\n",
            "  Attempting uninstall: spacy\n",
            "    Found existing installation: spacy 3.4.1\n",
            "    Uninstalling spacy-3.4.1:\n",
            "      Successfully uninstalled spacy-3.4.1\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "en-core-web-sm 3.4.0 requires spacy<3.5.0,>=3.4.0, but you have spacy 3.2.4 which is incompatible.\u001b[0m\n",
            "Successfully installed conllu-4.5.2 nmslib-2.1.1 pybind11-2.6.1 pydantic-1.8.2 pysbd-0.3.4 scispacy-0.5.0 spacy-3.2.4 thinc-8.0.17 typing-extensions-3.10.0.2\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "typing_extensions"
                ]
              }
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.0/en_core_sci_sm-0.5.0.tar.gz\n",
            "  Downloading https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.0/en_core_sci_sm-0.5.0.tar.gz (15.9 MB)\n",
            "\u001b[K     |████████████████████████████████| 15.9 MB 1.1 MB/s \n",
            "\u001b[?25hRequirement already satisfied: spacy<3.3.0,>=3.2.3 in /usr/local/lib/python3.7/dist-packages (from en-core-sci-sm==0.5.0) (3.2.4)\n",
            "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.0.3)\n",
            "Requirement already satisfied: srsly<3.0.0,>=2.4.1 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.4.4)\n",
            "Requirement already satisfied: wasabi<1.1.0,>=0.8.1 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (0.10.1)\n",
            "Requirement already satisfied: typer<0.5.0,>=0.3.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (0.4.2)\n",
            "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<1.9.0,>=1.7.4 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.8.2)\n",
            "Requirement already satisfied: typing-extensions<4.0.0.0,>=3.7.4 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.10.0.2)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (4.64.0)\n",
            "Requirement already satisfied: langcodes<4.0.0,>=3.2.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.3.0)\n",
            "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.0.8)\n",
            "Requirement already satisfied: blis<0.8.0,>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (0.7.8)\n",
            "Requirement already satisfied: numpy>=1.15.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.21.6)\n",
            "Requirement already satisfied: thinc<8.1.0,>=8.0.12 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (8.0.17)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (21.3)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (57.4.0)\n",
            "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.8 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.0.10)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.0.6)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.13.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.23.0)\n",
            "Requirement already satisfied: click<8.1.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (7.1.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.11.3)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.0.8)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.0.7)\n",
            "Requirement already satisfied: pathy>=0.3.5 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (0.6.2)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from catalogue<2.1.0,>=2.0.6->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.8.1)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging>=20.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.0.9)\n",
            "Requirement already satisfied: smart-open<6.0.0,>=5.2.1 in /usr/local/lib/python3.7/dist-packages (from pathy>=0.3.5->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (5.2.1)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2022.6.15)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.24.3)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.10)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.0.4)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.7/dist-packages (from jinja2->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.0.1)\n",
            "Building wheels for collected packages: en-core-sci-sm\n",
            "  Building wheel for en-core-sci-sm (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for en-core-sci-sm: filename=en_core_sci_sm-0.5.0-py3-none-any.whl size=15871046 sha256=c7f9d32d0f706fd2361d0503da57e60e1177d43fe68d15720269ff6342239d7a\n",
            "  Stored in directory: /root/.cache/pip/wheels/b1/b8/7d/d26a5894db99bf0cfb9add399889738de91a608e6664f82454\n",
            "Successfully built en-core-sci-sm\n",
            "Installing collected packages: en-core-sci-sm\n",
            "Successfully installed en-core-sci-sm-0.5.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install flair"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LRn7z0Osr7Rm",
        "outputId": "c16af17f-41a4-4e2d-8844-5f49ce22c248"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting flair\n",
            "  Downloading flair-0.11.3-py3-none-any.whl (401 kB)\n",
            "\u001b[K     |████████████████████████████████| 401 kB 6.1 MB/s \n",
            "\u001b[?25hCollecting wikipedia-api\n",
            "  Downloading Wikipedia-API-0.5.4.tar.gz (18 kB)\n",
            "Requirement already satisfied: gdown==4.4.0 in /usr/local/lib/python3.7/dist-packages (from flair) (4.4.0)\n",
            "Collecting janome\n",
            "  Downloading Janome-0.4.2-py2.py3-none-any.whl (19.7 MB)\n",
            "\u001b[K     |████████████████████████████████| 19.7 MB 1.5 MB/s \n",
            "\u001b[?25hRequirement already satisfied: torch!=1.8,>=1.5.0 in /usr/local/lib/python3.7/dist-packages (from flair) (1.12.1+cu113)\n",
            "Requirement already satisfied: gensim>=3.4.0 in /usr/local/lib/python3.7/dist-packages (from flair) (3.6.0)\n",
            "Collecting mpld3==0.3\n",
            "  Downloading mpld3-0.3.tar.gz (788 kB)\n",
            "\u001b[K     |████████████████████████████████| 788 kB 10.4 MB/s \n",
            "\u001b[?25hCollecting huggingface-hub\n",
            "  Downloading huggingface_hub-0.9.1-py3-none-any.whl (120 kB)\n",
            "\u001b[K     |████████████████████████████████| 120 kB 49.6 MB/s \n",
            "\u001b[?25hCollecting segtok>=1.5.7\n",
            "  Downloading segtok-1.5.11-py3-none-any.whl (24 kB)\n",
            "Requirement already satisfied: python-dateutil>=2.6.1 in /usr/local/lib/python3.7/dist-packages (from flair) (2.8.2)\n",
            "Collecting deprecated>=1.2.4\n",
            "  Downloading Deprecated-1.2.13-py2.py3-none-any.whl (9.6 kB)\n",
            "Collecting konoha<5.0.0,>=4.0.0\n",
            "  Downloading konoha-4.6.5-py3-none-any.whl (20 kB)\n",
            "Requirement already satisfied: conllu>=4.0 in /usr/local/lib/python3.7/dist-packages (from flair) (4.5.2)\n",
            "Requirement already satisfied: regex in /usr/local/lib/python3.7/dist-packages (from flair) (2022.6.2)\n",
            "Collecting bpemb>=0.3.2\n",
            "  Downloading bpemb-0.3.3-py3-none-any.whl (19 kB)\n",
            "Collecting ftfy\n",
            "  Downloading ftfy-6.1.1-py3-none-any.whl (53 kB)\n",
            "\u001b[K     |████████████████████████████████| 53 kB 1.8 MB/s \n",
            "\u001b[?25hRequirement already satisfied: more-itertools in /usr/local/lib/python3.7/dist-packages (from flair) (8.14.0)\n",
            "Requirement already satisfied: tabulate in /usr/local/lib/python3.7/dist-packages (from flair) (0.8.10)\n",
            "Collecting transformers>=4.0.0\n",
            "  Downloading transformers-4.21.2-py3-none-any.whl (4.7 MB)\n",
            "\u001b[K     |████████████████████████████████| 4.7 MB 26.4 MB/s \n",
            "\u001b[?25hRequirement already satisfied: matplotlib>=2.2.3 in /usr/local/lib/python3.7/dist-packages (from flair) (3.2.2)\n",
            "Collecting hyperopt>=0.2.7\n",
            "  Downloading hyperopt-0.2.7-py2.py3-none-any.whl (1.6 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.6 MB 45.4 MB/s \n",
            "\u001b[?25hCollecting sqlitedict>=1.6.0\n",
            "  Downloading sqlitedict-2.0.0.tar.gz (46 kB)\n",
            "\u001b[K     |████████████████████████████████| 46 kB 4.1 MB/s \n",
            "\u001b[?25hCollecting sentencepiece==0.1.95\n",
            "  Downloading sentencepiece-0.1.95-cp37-cp37m-manylinux2014_x86_64.whl (1.2 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.2 MB 28.9 MB/s \n",
            "\u001b[?25hCollecting pptree\n",
            "  Downloading pptree-3.1.tar.gz (3.0 kB)\n",
            "Collecting langdetect\n",
            "  Downloading langdetect-1.0.9.tar.gz (981 kB)\n",
            "\u001b[K     |████████████████████████████████| 981 kB 27.1 MB/s \n",
            "\u001b[?25hRequirement already satisfied: tqdm>=4.26.0 in /usr/local/lib/python3.7/dist-packages (from flair) (4.64.0)\n",
            "Requirement already satisfied: scikit-learn>=0.21.3 in /usr/local/lib/python3.7/dist-packages (from flair) (1.0.2)\n",
            "Requirement already satisfied: lxml in /usr/local/lib/python3.7/dist-packages (from flair) (4.9.1)\n",
            "Requirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.7/dist-packages (from gdown==4.4.0->flair) (4.6.3)\n",
            "Requirement already satisfied: requests[socks] in /usr/local/lib/python3.7/dist-packages (from gdown==4.4.0->flair) (2.23.0)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.7/dist-packages (from gdown==4.4.0->flair) (3.8.0)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from gdown==4.4.0->flair) (1.15.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from bpemb>=0.3.2->flair) (1.21.6)\n",
            "Requirement already satisfied: wrapt<2,>=1.10 in /usr/local/lib/python3.7/dist-packages (from deprecated>=1.2.4->flair) (1.14.1)\n",
            "Requirement already satisfied: smart-open>=1.2.1 in /usr/local/lib/python3.7/dist-packages (from gensim>=3.4.0->flair) (5.2.1)\n",
            "Requirement already satisfied: scipy>=0.18.1 in /usr/local/lib/python3.7/dist-packages (from gensim>=3.4.0->flair) (1.7.3)\n",
            "Collecting py4j\n",
            "  Downloading py4j-0.10.9.7-py2.py3-none-any.whl (200 kB)\n",
            "\u001b[K     |████████████████████████████████| 200 kB 43.9 MB/s \n",
            "\u001b[?25hRequirement already satisfied: future in /usr/local/lib/python3.7/dist-packages (from hyperopt>=0.2.7->flair) (0.16.0)\n",
            "Requirement already satisfied: cloudpickle in /usr/local/lib/python3.7/dist-packages (from hyperopt>=0.2.7->flair) (1.5.0)\n",
            "Requirement already satisfied: networkx>=2.2 in /usr/local/lib/python3.7/dist-packages (from hyperopt>=0.2.7->flair) (2.6.3)\n",
            "Collecting overrides<4.0.0,>=3.0.0\n",
            "  Downloading overrides-3.1.0.tar.gz (11 kB)\n",
            "Collecting importlib-metadata<4.0.0,>=3.7.0\n",
            "  Downloading importlib_metadata-3.10.1-py3-none-any.whl (14 kB)\n",
            "Collecting requests\n",
            "  Downloading requests-2.28.1-py3-none-any.whl (62 kB)\n",
            "\u001b[K     |████████████████████████████████| 62 kB 1.6 MB/s \n",
            "\u001b[?25hRequirement already satisfied: typing-extensions>=3.6.4 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata<4.0.0,>=3.7.0->konoha<5.0.0,>=4.0.0->flair) (3.10.0.2)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata<4.0.0,>=3.7.0->konoha<5.0.0,>=4.0.0->flair) (3.8.1)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=2.2.3->flair) (3.0.9)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=2.2.3->flair) (1.4.4)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=2.2.3->flair) (0.11.0)\n",
            "Requirement already satisfied: charset-normalizer<3,>=2 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (2.1.1)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (1.24.3)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (2.10)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (2022.6.15)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.21.3->flair) (1.1.0)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.21.3->flair) (3.1.0)\n",
            "Requirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.7/dist-packages (from transformers>=4.0.0->flair) (6.0)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from transformers>=4.0.0->flair) (21.3)\n",
            "Collecting tokenizers!=0.11.3,<0.13,>=0.11.1\n",
            "  Downloading tokenizers-0.12.1-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (6.6 MB)\n",
            "\u001b[K     |████████████████████████████████| 6.6 MB 29.4 MB/s \n",
            "\u001b[?25hRequirement already satisfied: wcwidth>=0.2.5 in /usr/local/lib/python3.7/dist-packages (from ftfy->flair) (0.2.5)\n",
            "Requirement already satisfied: PySocks!=1.5.7,>=1.5.6 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (1.7.1)\n",
            "Building wheels for collected packages: mpld3, overrides, sqlitedict, langdetect, pptree, wikipedia-api\n",
            "  Building wheel for mpld3 (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for mpld3: filename=mpld3-0.3-py3-none-any.whl size=116702 sha256=df62c6eb2b37a30c799fd61f7a7f53bf44043679c9a6aeb92c781092d43f5240\n",
            "  Stored in directory: /root/.cache/pip/wheels/26/70/6a/1c79e59951a41b4045497da187b2724f5659ca64033cf4548e\n",
            "  Building wheel for overrides (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for overrides: filename=overrides-3.1.0-py3-none-any.whl size=10187 sha256=9d40f144b7533d3f875f53ca4924a29277a249adce5e6da9b26cc5c08a7b0af4\n",
            "  Stored in directory: /root/.cache/pip/wheels/3a/0d/38/01a9bc6e20dcfaf0a6a7b552d03137558ba1c38aea47644682\n",
            "  Building wheel for sqlitedict (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for sqlitedict: filename=sqlitedict-2.0.0-py3-none-any.whl size=15736 sha256=3e47c1d354e05a648a22be33d5fc113d90db04648984e8270bc9f1d2d6c86368\n",
            "  Stored in directory: /root/.cache/pip/wheels/96/dd/2e/0ed4a25cb73fc30c7ea8d10b50acb7226175736067e40a7ea3\n",
            "  Building wheel for langdetect (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for langdetect: filename=langdetect-1.0.9-py3-none-any.whl size=993242 sha256=a0a6758119f453c6bc3455e0145e32364e45f142e7cbff787b2c409985abc3c2\n",
            "  Stored in directory: /root/.cache/pip/wheels/c5/96/8a/f90c59ed25d75e50a8c10a1b1c2d4c402e4dacfa87f3aff36a\n",
            "  Building wheel for pptree (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pptree: filename=pptree-3.1-py3-none-any.whl size=4629 sha256=e05575e0b67618db39bd9c7326d0b3de847fe1b63eef16a48fb64ce9e7c79204\n",
            "  Stored in directory: /root/.cache/pip/wheels/9e/e8/7d/a9c3c19b4722608a0d8b05a38c36bc3f230c43becd2a46794b\n",
            "  Building wheel for wikipedia-api (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for wikipedia-api: filename=Wikipedia_API-0.5.4-py3-none-any.whl size=13477 sha256=0916b0cda94b8e9d90780b133cbc8ea282451454b8cf45a0438b6d8c82a42cf5\n",
            "  Stored in directory: /root/.cache/pip/wheels/d3/24/56/58ba93cf78be162451144e7a9889603f437976ef1ae7013d04\n",
            "Successfully built mpld3 overrides sqlitedict langdetect pptree wikipedia-api\n",
            "Installing collected packages: requests, importlib-metadata, tokenizers, sentencepiece, py4j, overrides, huggingface-hub, wikipedia-api, transformers, sqlitedict, segtok, pptree, mpld3, langdetect, konoha, janome, hyperopt, ftfy, deprecated, bpemb, flair\n",
            "  Attempting uninstall: requests\n",
            "    Found existing installation: requests 2.23.0\n",
            "    Uninstalling requests-2.23.0:\n",
            "      Successfully uninstalled requests-2.23.0\n",
            "  Attempting uninstall: importlib-metadata\n",
            "    Found existing installation: importlib-metadata 4.12.0\n",
            "    Uninstalling importlib-metadata-4.12.0:\n",
            "      Successfully uninstalled importlib-metadata-4.12.0\n",
            "  Attempting uninstall: hyperopt\n",
            "    Found existing installation: hyperopt 0.1.2\n",
            "    Uninstalling hyperopt-0.1.2:\n",
            "      Successfully uninstalled hyperopt-0.1.2\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "markdown 3.4.1 requires importlib-metadata>=4.4; python_version < \"3.10\", but you have importlib-metadata 3.10.1 which is incompatible.\n",
            "gym 0.25.2 requires importlib-metadata>=4.8.0; python_version < \"3.10\", but you have importlib-metadata 3.10.1 which is incompatible.\n",
            "en-core-web-sm 3.4.0 requires spacy<3.5.0,>=3.4.0, but you have spacy 3.2.4 which is incompatible.\u001b[0m\n",
            "Successfully installed bpemb-0.3.3 deprecated-1.2.13 flair-0.11.3 ftfy-6.1.1 huggingface-hub-0.9.1 hyperopt-0.2.7 importlib-metadata-3.10.1 janome-0.4.2 konoha-4.6.5 langdetect-1.0.9 mpld3-0.3 overrides-3.1.0 pptree-3.1 py4j-0.10.9.7 requests-2.28.1 segtok-1.5.11 sentencepiece-0.1.95 sqlitedict-2.0.0 tokenizers-0.12.1 transformers-4.21.2 wikipedia-api-0.5.4\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "2gtZwz-ksFpO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import requests\n",
        "from bs4 import BeautifulSoup\n",
        "#from pubmed_lookup import PubMedLookup\n",
        "#from pubmed_lookup import Publication\n",
        "from joblib import Parallel, delayed"
      ],
      "metadata": {
        "id": "PZ6Bfiyxsluz"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')\n",
        "%cd /content/gdrive/MyDrive/Github/"
      ],
      "metadata": {
        "id": "zC5l8iHFd9uT",
        "outputId": "c8f436f1-4fee-4e9d-bc0c-3117e16b3736",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/gdrive\n",
            "/content/gdrive/MyDrive/Github\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Connecting to the GIT repository\n",
        "! git config --global user.email \"plarmande@gmail.com\"\n",
        "! git config --global user.name \"pierrelarmande\"\n",
        "\n",
        "username = 'pierrelarmande'\n",
        "repo = 'OryzaGP'\n",
        "# To create a token : https://github.com/settings/tokens\n",
        "git_token = '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'\n",
        "\n",
        "url = \"https://\"+git_token+\"@github.com/\"+username+\"/\"+repo+\".git\""
      ],
      "metadata": {
        "id": "MIH7MBMLeDz3"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Clone if doesn't already exist\n",
        "#!git clone {url}\n",
        "%pwd\n",
        "%cd $repo\n",
        "#%cd $repo\n",
        "#! git pull"
      ],
      "metadata": {
        "id": "rjPAm2MOfm9T",
        "outputId": "4e5f5d8d-e041-4cc3-a130-94e2b2cd54d8",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/gdrive/MyDrive/Github/OryzaGP\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Get data\n",
        "reference_file = r'https://shigen.nig.ac.jp/rice/oryzabase/reference/download'\n",
        "gene_file = r'https://shigen.nig.ac.jp/rice/oryzabase/gene/download?classtag=GENE_EN_LIST'\n",
        "reference_file = 'https://shigen.nig.ac.jp/rice/oryzabase/reference/download'\n",
        "base_url = \"https://shigen.nig.ac.jp/rice/oryzabase/gene/detail/\"\n",
        "#page = requests.get(base_url + \"477\")\n",
        "output_file  = 'reference_cleaned.txt'\n",
        "pub_mappings_geneid = 'pub_mappings_genid.txt'\n",
        "abstracts_file = 'output/abstracts.tsv'\n",
        "\n",
        "#iric = pd.DataFrame(pd.read_pickle('output/iric_oryzabase_tf_rapdb_annotation.pkl'))\n",
        "!ls output"
      ],
      "metadata": {
        "id": "NA3H-miMftVJ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d0250b47-77db-4fdb-bb0b-518498a70d17"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "abstracts.csv  abstracts.tsv  abstracts.xml  info.md\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pd_gene = pd.read_csv(gene_file, sep='\\t', delimiter=None, index_col= 0, dtype='str', encoding = \"ISO-8859-1\", error_bad_lines=False, skip_blank_lines=True)"
      ],
      "metadata": {
        "id": "DGWcbrPC6lVr",
        "outputId": "0adec16f-19bd-4d52-8ea6-e2ea0aef327a",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.\n",
            "\n",
            "\n",
            "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "o_lr8Kpw7Vly"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "array = pd.read_csv(reference_file, sep='\\t', delimiter=None, index_col= 0, dtype='str', encoding = \"ISO-8859-1\", error_bad_lines=False, skip_blank_lines=True)"
      ],
      "metadata": {
        "id": "DfqgIBTWlt97",
        "outputId": "4595fcb8-097d-4b53-d1f0-51e64f407306",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.\n",
            "\n",
            "\n",
            "  exec(code_obj, self.user_global_ns, self.user_ns)\n",
            "b'Skipping line 11636: expected 10 fields, saw 12\\n'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "array.head()"
      ],
      "metadata": {
        "id": "iovO4pnklx8e",
        "outputId": "598a3460-85db-432f-aa20-1a77cf778241",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 368
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "             PubMedId                                             Author  \\\n",
              "Reference_id                                                               \n",
              "1                 NaN  Abenes, M.L.P., R.E. Tabien, S.R. McCouch, R. ...   \n",
              "2                 NaN                                         Adachi, T.   \n",
              "3             8103599                          Ahn, S. and S.D. Tanksley   \n",
              "4                 NaN  Ahn, S.N., C.N. Bollich, A.M. McClung and S.D....   \n",
              "5             7903411  Ahn, S., J.A. Anderson, N.E. Sorrels and S.D. ...   \n",
              "\n",
              "                                                          Title  \\\n",
              "Reference_id                                                      \n",
              "1             Orientation and integration of the classical a...   \n",
              "2                 Dwarfness and leafy head newly found in rice.   \n",
              "3             Comparative linkage maps of the rice and maize...   \n",
              "4             RFLP analysis of genomic regions associated wi...   \n",
              "5             Homologous relationships of rice, wheat and ma...   \n",
              "\n",
              "                                    Journal Volume       Pages  Year  \\\n",
              "Reference_id                                                           \n",
              "1                                 Euphytica     76      81-87.  1994   \n",
              "2                            Agric. & Hort.  10(4)  1048-1050.  1935   \n",
              "3             Proc. Natl. Acad. Sci. U.S.A.     90  7980-7984.  1993   \n",
              "4                       Theor. Appl. Genet.     87      27-32.  1993   \n",
              "5                          Mol. Gen. Genet.    241    483-490.  1993   \n",
              "\n",
              "                CGSNL Gene Symbol  \\\n",
              "Reference_id                        \n",
              "1             LA,D27,Z2,XA21,ADH1   \n",
              "2                             LHD   \n",
              "3                             NaN   \n",
              "4                SK2(T)(SCL, FGR)   \n",
              "5                             NaN   \n",
              "\n",
              "                                              Gene Name Synonym  \n",
              "Reference_id                                                     \n",
              "1             la,d27(d t), dt, dwf23, d27, OsD27,z2,Xa21*, X...  \n",
              "2                                                           lhd  \n",
              "3                                                           NaN  \n",
              "4             sk2(t)(scl, fgr), fgr, scl, BAD2, BADH2, sk2, ...  \n",
              "5                                                           NaN  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-a8115f11-599f-48fa-aca7-ce2384517a80\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>PubMedId</th>\n",
              "      <th>Author</th>\n",
              "      <th>Title</th>\n",
              "      <th>Journal</th>\n",
              "      <th>Volume</th>\n",
              "      <th>Pages</th>\n",
              "      <th>Year</th>\n",
              "      <th>CGSNL Gene Symbol</th>\n",
              "      <th>Gene Name Synonym</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Reference_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Abenes, M.L.P., R.E. Tabien, S.R. McCouch, R. ...</td>\n",
              "      <td>Orientation and integration of the classical a...</td>\n",
              "      <td>Euphytica</td>\n",
              "      <td>76</td>\n",
              "      <td>81-87.</td>\n",
              "      <td>1994</td>\n",
              "      <td>LA,D27,Z2,XA21,ADH1</td>\n",
              "      <td>la,d27(d t), dt, dwf23, d27, OsD27,z2,Xa21*, X...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Adachi, T.</td>\n",
              "      <td>Dwarfness and leafy head newly found in rice.</td>\n",
              "      <td>Agric. &amp; Hort.</td>\n",
              "      <td>10(4)</td>\n",
              "      <td>1048-1050.</td>\n",
              "      <td>1935</td>\n",
              "      <td>LHD</td>\n",
              "      <td>lhd</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>8103599</td>\n",
              "      <td>Ahn, S. and S.D. Tanksley</td>\n",
              "      <td>Comparative linkage maps of the rice and maize...</td>\n",
              "      <td>Proc. Natl. Acad. Sci. U.S.A.</td>\n",
              "      <td>90</td>\n",
              "      <td>7980-7984.</td>\n",
              "      <td>1993</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Ahn, S.N., C.N. Bollich, A.M. McClung and S.D....</td>\n",
              "      <td>RFLP analysis of genomic regions associated wi...</td>\n",
              "      <td>Theor. Appl. Genet.</td>\n",
              "      <td>87</td>\n",
              "      <td>27-32.</td>\n",
              "      <td>1993</td>\n",
              "      <td>SK2(T)(SCL, FGR)</td>\n",
              "      <td>sk2(t)(scl, fgr), fgr, scl, BAD2, BADH2, sk2, ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>7903411</td>\n",
              "      <td>Ahn, S., J.A. Anderson, N.E. Sorrels and S.D. ...</td>\n",
              "      <td>Homologous relationships of rice, wheat and ma...</td>\n",
              "      <td>Mol. Gen. Genet.</td>\n",
              "      <td>241</td>\n",
              "      <td>483-490.</td>\n",
              "      <td>1993</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-a8115f11-599f-48fa-aca7-ce2384517a80')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-a8115f11-599f-48fa-aca7-ce2384517a80 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-a8115f11-599f-48fa-aca7-ce2384517a80');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "array.count()"
      ],
      "metadata": {
        "id": "-lAoncTFl6U7",
        "outputId": "f88a6433-f573-4169-ab1f-365088aa9880",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PubMedId             19898\n",
              "Author               21936\n",
              "Title                21944\n",
              "Journal              21930\n",
              "Volume               20053\n",
              "Pages                19300\n",
              "Year                 21913\n",
              "CGSNL Gene Symbol     8645\n",
              "Gene Name Synonym     8577\n",
              "dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#**Clean up the Reference dataset**"
      ],
      "metadata": {
        "id": "NYNvUcStmJfd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "result = array.dropna(subset=['PubMedId'])"
      ],
      "metadata": {
        "id": "p0PIqTjamd8b"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "result.count()"
      ],
      "metadata": {
        "id": "b-7mBLfhmiQg",
        "outputId": "49927ffd-ccb5-463e-aba3-97d9131e6d20",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PubMedId             19898\n",
              "Author               19879\n",
              "Title                19886\n",
              "Journal              19872\n",
              "Volume               18153\n",
              "Pages                17252\n",
              "Year                 19855\n",
              "CGSNL Gene Symbol     7174\n",
              "Gene Name Synonym     7125\n",
              "dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# donnees.drop(['Unnamed: 15', 'Unnamed: 16'], axis=1,inplace=True)\n",
        "final_result = result.drop(['Author','Title','Journal','Volume','Pages','Year'], axis=1)"
      ],
      "metadata": {
        "id": "1INmsLfamr5L"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "final_result.head()"
      ],
      "metadata": {
        "id": "BfYUHeRzm_yD",
        "outputId": "e7b6753e-86fd-4248-fbff-8addf4751afe",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "              PubMedId CGSNL Gene Symbol  \\\n",
              "Reference_id                               \n",
              "3              8103599               NaN   \n",
              "5              7903411               NaN   \n",
              "42             7247938              ACP1   \n",
              "45            17248812           SD2,SD3   \n",
              "113           18470071          EF,BPH10   \n",
              "\n",
              "                                    Gene Name Synonym  \n",
              "Reference_id                                           \n",
              "3                                                 NaN  \n",
              "5                                                 NaN  \n",
              "42                                               Acp1  \n",
              "45                                            sd2,sd3  \n",
              "113           Ef(t)*, ef(t)*, Ef, ef,Bph10(t)*, Bph10  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-feaf9a56-ca44-4a7f-97df-ff9f5edc707f\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>PubMedId</th>\n",
              "      <th>CGSNL Gene Symbol</th>\n",
              "      <th>Gene Name Synonym</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Reference_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>8103599</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>7903411</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>7247938</td>\n",
              "      <td>ACP1</td>\n",
              "      <td>Acp1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>17248812</td>\n",
              "      <td>SD2,SD3</td>\n",
              "      <td>sd2,sd3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>113</th>\n",
              "      <td>18470071</td>\n",
              "      <td>EF,BPH10</td>\n",
              "      <td>Ef(t)*, ef(t)*, Ef, ef,Bph10(t)*, Bph10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-feaf9a56-ca44-4a7f-97df-ff9f5edc707f')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-feaf9a56-ca44-4a7f-97df-ff9f5edc707f button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-feaf9a56-ca44-4a7f-97df-ff9f5edc707f');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final_result=final_result.fillna('_')"
      ],
      "metadata": {
        "id": "AUF1lH7qmt9L"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(len(final_result[final_result['CGSNL Gene Symbol'] == '_']))"
      ],
      "metadata": {
        "id": "K33Ah_0Smzvn",
        "outputId": "f5263435-34f0-41a3-d508-5047c7577c78",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "13081\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final_result.astype({'PubMedId': 'int32'}).dtypes"
      ],
      "metadata": {
        "id": "eg-3qbTLT4wL",
        "outputId": "b7af2e40-a523-47f7-eae4-61cb7e34c3bb",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PubMedId              int32\n",
              "CGSNL Gene Symbol    object\n",
              "Gene Name Synonym    object\n",
              "dtype: object"
            ]
          },
          "metadata": {},
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final_result['PubMedId'] = pd.to_numeric(final_result['PubMedId'])"
      ],
      "metadata": {
        "id": "EEsPhh0hWmyO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "final_result['PubMedId'].value_counts()[0]"
      ],
      "metadata": {
        "id": "3TRIpbbZUecB",
        "outputId": "42a186e3-a355-408d-8c88-4fd1b9ba3cd4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "2116"
            ]
          },
          "metadata": {},
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pmids_list = final_result['PubMedId'].to_numpy()"
      ],
      "metadata": {
        "id": "Ps7M1dGh-rd1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "final_result.dtypes"
      ],
      "metadata": {
        "id": "TAgu0kQIV1oD",
        "outputId": "3c4f447c-d0e5-49d0-8411-672efda0a4cf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PubMedId              int64\n",
              "CGSNL Gene Symbol    object\n",
              "Gene Name Synonym    object\n",
              "dtype: object"
            ]
          },
          "metadata": {},
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pmids_list"
      ],
      "metadata": {
        "id": "k6uovoncS_W-",
        "outputId": "300cfc69-fe4b-4261-88c2-c7f57ec17f93",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([ 8103599,  7903411,  7247938, ..., 35642508, 35812946, 35832225])"
            ]
          },
          "metadata": {},
          "execution_count": 23
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pmids_set = set(pmids_list)"
      ],
      "metadata": {
        "id": "b5HXzTfQXLGz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pmids = list(pmids_set)"
      ],
      "metadata": {
        "id": "NwCUHb85XOad"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pmids[1:10]"
      ],
      "metadata": {
        "id": "psSJ5n1S1H0a",
        "outputId": "1f916358-6ff6-463e-b028-20b93fa7a069",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[31424513,\n",
              " 12582914,\n",
              " 22020099,\n",
              " 12582916,\n",
              " 25821195,\n",
              " 20938764,\n",
              " 25755661,\n",
              " 18612237,\n",
              " 31621135]"
            ]
          },
          "metadata": {},
          "execution_count": 26
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Exporting mapping Gene_ID and Gene Name"
      ],
      "metadata": {
        "id": "VgU8tfUK8E1V"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "pd_gene.head(3)"
      ],
      "metadata": {
        "id": "x2gqoBbk8NDY",
        "outputId": "2bb39f1d-9ec6-40a1-b4ed-a28e310cccb5",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 569
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                CGSNL Gene Symbol                  Gene symbol synonym(s)  \\\n",
              "Trait Gene Id                                                               \n",
              "1                     [CMS-54257]               [cms-54257]*, [cms-54257]   \n",
              "2              [CMS-AK]([CMS-JP])  [cms-ak]([cms-jp]), [cms-jp], [cms-ak]   \n",
              "3                       [CMS-ARC]   [cms-ARC]*, [cms-ARC] [mt], [cms-ARC]   \n",
              "\n",
              "                                CGSNL Gene Name  \\\n",
              "Trait Gene Id                                     \n",
              "1              CYTOPLASMIC MALE STERILITY 54257   \n",
              "2                 CYTOPLASMIC MALE STERILITY AK   \n",
              "3                CYTOPLASMIC MALE STERILITY ARC   \n",
              "\n",
              "                                            Gene name synonym(s) Protein Name  \\\n",
              "Trait Gene Id                                                                   \n",
              "1              Cytoplasmic mutant induced by somaclonal varia...          NaN   \n",
              "2                        Akebono' cytoplasm, 'Akebono' cytoplasm          NaN   \n",
              "3                 ARC13829-16 cytoplasm, `ARC13829-26' cytoplasm          NaN   \n",
              "\n",
              "              Allele Chromosome No.  \\\n",
              "Trait Gene Id                         \n",
              "1                NaN            NaN   \n",
              "2                NaN            NaN   \n",
              "3                NaN            NaN   \n",
              "\n",
              "                                                     Explanation  \\\n",
              "Trait Gene Id                                                      \n",
              "1                                            PO:0009066; anther.   \n",
              "2                                            PO:0009066; anther.   \n",
              "3              PO:0006452; microspore (sensu Poaceae) ; PO:00...   \n",
              "\n",
              "                                                     Trait Class RAP ID  \\\n",
              "Trait Gene Id                                                             \n",
              "1               Reproductive organ - Pollination, fertilizati...    NaN   \n",
              "2               Reproductive organ - Pollination, fertilizati...    NaN   \n",
              "3               Reproductive organ - Pollination, fertilizati...    NaN   \n",
              "\n",
              "              MSU ID  Gramene ID  Arm Locate(cM)  \\\n",
              "Trait Gene Id                                      \n",
              "1                NaN  GR:0060167  NaN        NaN   \n",
              "2                NaN  GR:0060144  NaN        NaN   \n",
              "3                NaN  GR:0060145  NaN        NaN   \n",
              "\n",
              "                                                   Gene Ontology  \\\n",
              "Trait Gene Id                                                      \n",
              "1              GO:0000001 - mitochondrion inheritance, GO:000...   \n",
              "2              GO:0000001 - mitochondrion inheritance, GO:000...   \n",
              "3              GO:0007275 - multicellular organismal development   \n",
              "\n",
              "                                                  Trait Ontology  \\\n",
              "Trait Gene Id                                                      \n",
              "1              TO:0000232 - cytoplasmic male sterility (sensu...   \n",
              "2              TO:0000232 - cytoplasmic male sterility (sensu...   \n",
              "3              TO:0000232 - cytoplasmic male sterility (sensu...   \n",
              "\n",
              "                                                  Plant Ontology  \n",
              "Trait Gene Id                                                     \n",
              "1                                           PO:0009066 - anther   \n",
              "2                                           PO:0009066 - anther   \n",
              "3              PO:0009082 - spikelet floret , PO:0020048 - mi...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-e460b747-3705-4506-a6e4-ad6e8222852d\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>CGSNL Gene Symbol</th>\n",
              "      <th>Gene symbol synonym(s)</th>\n",
              "      <th>CGSNL Gene Name</th>\n",
              "      <th>Gene name synonym(s)</th>\n",
              "      <th>Protein Name</th>\n",
              "      <th>Allele</th>\n",
              "      <th>Chromosome No.</th>\n",
              "      <th>Explanation</th>\n",
              "      <th>Trait Class</th>\n",
              "      <th>RAP ID</th>\n",
              "      <th>MSU ID</th>\n",
              "      <th>Gramene ID</th>\n",
              "      <th>Arm</th>\n",
              "      <th>Locate(cM)</th>\n",
              "      <th>Gene Ontology</th>\n",
              "      <th>Trait Ontology</th>\n",
              "      <th>Plant Ontology</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Trait Gene Id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>[CMS-54257]</td>\n",
              "      <td>[cms-54257]*, [cms-54257]</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY 54257</td>\n",
              "      <td>Cytoplasmic mutant induced by somaclonal varia...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>PO:0009066; anther.</td>\n",
              "      <td>Reproductive organ - Pollination, fertilizati...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GR:0060167</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GO:0000001 - mitochondrion inheritance, GO:000...</td>\n",
              "      <td>TO:0000232 - cytoplasmic male sterility (sensu...</td>\n",
              "      <td>PO:0009066 - anther</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>[CMS-AK]([CMS-JP])</td>\n",
              "      <td>[cms-ak]([cms-jp]), [cms-jp], [cms-ak]</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY AK</td>\n",
              "      <td>Akebono' cytoplasm, 'Akebono' cytoplasm</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>PO:0009066; anther.</td>\n",
              "      <td>Reproductive organ - Pollination, fertilizati...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GR:0060144</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GO:0000001 - mitochondrion inheritance, GO:000...</td>\n",
              "      <td>TO:0000232 - cytoplasmic male sterility (sensu...</td>\n",
              "      <td>PO:0009066 - anther</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>[CMS-ARC]</td>\n",
              "      <td>[cms-ARC]*, [cms-ARC] [mt], [cms-ARC]</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY ARC</td>\n",
              "      <td>ARC13829-16 cytoplasm, `ARC13829-26' cytoplasm</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>PO:0006452; microspore (sensu Poaceae) ; PO:00...</td>\n",
              "      <td>Reproductive organ - Pollination, fertilizati...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GR:0060145</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>GO:0007275 - multicellular organismal development</td>\n",
              "      <td>TO:0000232 - cytoplasmic male sterility (sensu...</td>\n",
              "      <td>PO:0009082 - spikelet floret , PO:0020048 - mi...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e460b747-3705-4506-a6e4-ad6e8222852d')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-e460b747-3705-4506-a6e4-ad6e8222852d button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-e460b747-3705-4506-a6e4-ad6e8222852d');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 30
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pd_gene.columns = [str(x).replace('(','_').replace(\n",
        "                     ' ','_').replace(\n",
        "                     '.','').replace(\n",
        "                    ')','') for x in list(pd_gene.axes[1])]"
      ],
      "metadata": {
        "id": "lnqGisv58XD1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pd_gene.columns"
      ],
      "metadata": {
        "id": "FL22y0XG8kSC",
        "outputId": "d833943e-68db-4a79-afc0-2055dbbb7f86",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['CGSNL_Gene_Symbol', 'Gene_symbol_synonym_s', 'CGSNL_Gene_Name',\n",
              "       'Gene_name_synonym_s', 'Protein_Name', 'Allele', 'Chromosome_No',\n",
              "       'Explanation', 'Trait_Class', 'RAP_ID', 'MSU_ID', 'Gramene_ID', 'Arm',\n",
              "       'Locate_cM', 'Gene_Ontology', 'Trait_Ontology', 'Plant_Ontology'],\n",
              "      dtype='object')"
            ]
          },
          "metadata": {},
          "execution_count": 32
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pd_mapping = pd_gene[['CGSNL_Gene_Name', 'Gene_name_synonym_s','RAP_ID', 'MSU_ID']]"
      ],
      "metadata": {
        "id": "uVDY82e68wCO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pd_mapping = pd_mapping.reset_index()"
      ],
      "metadata": {
        "id": "kbyHvrLz9HD0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pd_mapping.head()"
      ],
      "metadata": {
        "id": "OpDh-S7u_XS8",
        "outputId": "a7ef25f7-d9cb-4d45-b7c0-836c9ceec0f6",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "   Trait Gene Id                   CGSNL_Gene_Name  \\\n",
              "0              1  CYTOPLASMIC MALE STERILITY 54257   \n",
              "1              2     CYTOPLASMIC MALE STERILITY AK   \n",
              "2              3    CYTOPLASMIC MALE STERILITY ARC   \n",
              "3              4     CYTOPLASMIC MALE STERILITY BO   \n",
              "4              5     CYTOPLASMIC MALE STERILITY CW   \n",
              "\n",
              "                                 Gene_name_synonym_s RAP_ID MSU_ID  \n",
              "0  Cytoplasmic mutant induced by somaclonal varia...   None   None  \n",
              "1            Akebono' cytoplasm, 'Akebono' cytoplasm   None   None  \n",
              "2     ARC13829-16 cytoplasm, `ARC13829-26' cytoplasm   None   None  \n",
              "3  Chinsurah boro II' cytoplasm, `Chinsurah boro ...   None   None  \n",
              "4                        Chinese wild rice cytoplasm   None   None  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-a6dbbcf5-b717-47df-b213-aa16f9b63418\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Trait Gene Id</th>\n",
              "      <th>CGSNL_Gene_Name</th>\n",
              "      <th>Gene_name_synonym_s</th>\n",
              "      <th>RAP_ID</th>\n",
              "      <th>MSU_ID</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY 54257</td>\n",
              "      <td>Cytoplasmic mutant induced by somaclonal varia...</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>2</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY AK</td>\n",
              "      <td>Akebono' cytoplasm, 'Akebono' cytoplasm</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>3</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY ARC</td>\n",
              "      <td>ARC13829-16 cytoplasm, `ARC13829-26' cytoplasm</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>4</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY BO</td>\n",
              "      <td>Chinsurah boro II' cytoplasm, `Chinsurah boro ...</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>5</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY CW</td>\n",
              "      <td>Chinese wild rice cytoplasm</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-a6dbbcf5-b717-47df-b213-aa16f9b63418')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-a6dbbcf5-b717-47df-b213-aa16f9b63418 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-a6dbbcf5-b717-47df-b213-aa16f9b63418');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 40
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pd_mapping=pd_mapping.fillna('None')"
      ],
      "metadata": {
        "id": "_zv5uplz9ta6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "mapping_list=[]\n",
        "rap_lit=[]\n",
        "msu_list=[]\n",
        "for row in pd_mapping.to_numpy():\n",
        "  if row[1] != 'None' and row[1] != '_':\n",
        "    name = row[1].upper()\n",
        "    mapping_list.append([row[0],name])\n",
        "  elif row[2] != 'None' and row[2] != '_':\n",
        "    if ',' in row[2]:\n",
        "      name = row[2].split(',')[0]\n",
        "      name = name.upper()\n",
        "      mapping_list.append([row[0],name])\n",
        "    else:\n",
        "      name = row[2].upper()\n",
        "      mapping_list.append([row[0],name])\n",
        "\n",
        "df = pd.DataFrame(mapping_list, columns =['Gene_ID', 'Name'])"
      ],
      "metadata": {
        "id": "14eBuOOK90f3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df['Gene_ID'] = 'http://identifiers.org/oryzabase.gene/' + df['Gene_ID'].astype(str)"
      ],
      "metadata": {
        "id": "rZtjjRkIBNkN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df.head()"
      ],
      "metadata": {
        "id": "vDb0qfGw-g4z",
        "outputId": "06cf10e0-aff0-4f95-8b1a-bb588ec12fa4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                   Gene_ID                              Name\n",
              "0  http://identifiers.org/oryzabase.gene/1  CYTOPLASMIC MALE STERILITY 54257\n",
              "1  http://identifiers.org/oryzabase.gene/2     CYTOPLASMIC MALE STERILITY AK\n",
              "2  http://identifiers.org/oryzabase.gene/3    CYTOPLASMIC MALE STERILITY ARC\n",
              "3  http://identifiers.org/oryzabase.gene/4     CYTOPLASMIC MALE STERILITY BO\n",
              "4  http://identifiers.org/oryzabase.gene/5     CYTOPLASMIC MALE STERILITY CW"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-369e490c-a4c7-4f96-952c-dc16a3d9d42d\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>http://identifiers.org/oryzabase.gene/1</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY 54257</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>http://identifiers.org/oryzabase.gene/2</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY AK</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>http://identifiers.org/oryzabase.gene/3</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY ARC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>http://identifiers.org/oryzabase.gene/4</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY BO</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>http://identifiers.org/oryzabase.gene/5</td>\n",
              "      <td>CYTOPLASMIC MALE STERILITY CW</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-369e490c-a4c7-4f96-952c-dc16a3d9d42d')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-369e490c-a4c7-4f96-952c-dc16a3d9d42d button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-369e490c-a4c7-4f96-952c-dc16a3d9d42d');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 53
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from pandas._libs import index\n",
        "df.to_csv('pubAnnotation.txt', sep='\\t',columns=['Name','Gene_ID'],index=False)"
      ],
      "metadata": {
        "id": "4PRxl-f_JIIl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Retrieving Title and Abstracts from PubMed"
      ],
      "metadata": {
        "id": "YS9bfR8D76Ju"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import Entrez\n",
        "import time\n",
        "Entrez.email = 'pierre.larmande@ird.fr'\n",
        "#pmids = [29090559 29058482 28991880 28984387 28862677 28804631 28801717 28770950 28768831 28707064 28701466 28685492 28623948 28551248]\n",
        "handle = Entrez.efetch(db=\"pubmed\", id=pmids[1:10], retmode=\"xml\")\n",
        "records = Entrez.read(handle)\n",
        "#abstracts = [pubmed_article['MedlineCitation']['Article']['Abstract']['AbstractText'][0]  if 'Abstract' in pubmed_article['MedlineCitation']['Article'].keys() else pubmed_article['MedlineCitation']['Article']['ArticleTitle']  for pubmed_article in records['PubmedArticle']]\n",
        "#abstract_dict = dict(zip(pmids, abstracts))\n",
        "#print abstract_dict\n"
      ],
      "metadata": {
        "id": "C6rYjjq4SzaI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "output_writer = open(abstracts_file, \"w\")"
      ],
      "metadata": {
        "id": "OdfPuJ98w-c9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#abstracts = [pubmed_article['MedlineCitation']['Article']['Abstract']['AbstractText'][0]  if 'Abstract' in pubmed_article['MedlineCitation']['Article'].keys() else pubmed_article['MedlineCitation']['Article']['ArticleTitle']  for pubmed_article in records['PubmedArticle']]\n",
        "#abstract_dict = dict(zip(pmids, abstracts))\n",
        "#print(abstract_dict)\n",
        "for pubmed_article in records['PubmedArticle']:\n",
        "  pmid = pubmed_article['MedlineCitation']['PMID'][:]\n",
        "  title = pubmed_article['MedlineCitation']['Article']['ArticleTitle']\n",
        "  abstract = pubmed_article['MedlineCitation']['Article']['Abstract']['AbstractText'][0]  if 'Abstract' in pubmed_article['MedlineCitation']['Article'].keys() else ''\n",
        "  #print(pmid)\n",
        "  output_writer.write(pmid + '\\t' + title + '\\t' + abstract +'\\n')\n",
        "\n",
        "handle.close()\n",
        "output_writer.close"
      ],
      "metadata": {
        "id": "Cni7JTkywopv",
        "outputId": "f47521e3-f2d0-4a4f-f290-7d76cc827682",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<function TextIOWrapper.close>"
            ]
          },
          "metadata": {},
          "execution_count": 52
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "scbTFDg4rviy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# NER Tagging"
      ],
      "metadata": {
        "id": "4NEG_vrz416q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from flair.data import Sentence\n",
        "from flair.models import MultiTagger\n",
        "from flair.tokenization import SciSpacyTokenizer\n",
        "from flair.tokenization import SciSpacySentenceSplitter\n",
        "import en_core_sci_sm\n",
        "nlp = en_core_sci_sm.load()"
      ],
      "metadata": {
        "id": "QOqlZ6_CrxPs"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# load biomedical tagger\n",
        "tagger = MultiTagger.load(\"hunflair\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sSeKKFcst9Fr",
        "outputId": "13824ce7-2c9a-493c-e516-9622dcce435d"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:06:43,427 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_smallish_models/cellline/hunflair-celline-v1.0.pt not found in cache, downloading to /tmp/tmpogwi297g\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104886372/1104886372 [00:51<00:00, 21637969.39B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:07:34,889 copying /tmp/tmpogwi297g to cache at /root/.flair/models/hunflair-celline-v1.0.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:07:39,403 removing temp file /tmp/tmpogwi297g\n",
            "2022-08-31 15:07:39,605 loading file /root/.flair/models/hunflair-celline-v1.0.pt\n",
            "2022-08-31 15:07:48,019 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-CellLine, B-CellLine, I-CellLine, E-CellLine, <START>, <STOP>\n",
            "2022-08-31 15:07:48,949 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_allcorpus_models/huner-chemical/hunflair-chemical-full-v1.0.pt not found in cache, downloading to /tmp/tmp2c7dztrm\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104888221/1104888221 [00:49<00:00, 22535934.72B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:08:38,382 copying /tmp/tmp2c7dztrm to cache at /root/.flair/models/hunflair-chemical-full-v1.0.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:08:42,460 removing temp file /tmp/tmp2c7dztrm\n",
            "2022-08-31 15:08:42,605 loading file /root/.flair/models/hunflair-chemical-full-v1.0.pt\n",
            "2022-08-31 15:08:48,154 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Chemical, B-Chemical, I-Chemical, E-Chemical, <START>, <STOP>\n",
            "2022-08-31 15:08:48,955 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_allcorpus_models/huner-disease/hunflair-disease-full-v1.0.pt not found in cache, downloading to /tmp/tmpcap2r1gk\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104886193/1104886193 [00:50<00:00, 21816867.81B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:09:40,027 copying /tmp/tmpcap2r1gk to cache at /root/.flair/models/hunflair-disease-full-v1.0.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:09:44,115 removing temp file /tmp/tmpcap2r1gk\n",
            "2022-08-31 15:09:44,261 loading file /root/.flair/models/hunflair-disease-full-v1.0.pt\n",
            "2022-08-31 15:09:49,977 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, B-Disease, E-Disease, I-Disease, S-Disease, <START>, <STOP>\n",
            "2022-08-31 15:09:51,000 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_allcorpus_models/huner-gene/hunflair-gene-full-v1.0.pt not found in cache, downloading to /tmp/tmpgh2idzlx\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104887211/1104887211 [00:48<00:00, 22623942.32B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:10:40,234 copying /tmp/tmpgh2idzlx to cache at /root/.flair/models/hunflair-gene-full-v1.0.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:10:44,267 removing temp file /tmp/tmpgh2idzlx\n",
            "2022-08-31 15:10:44,426 loading file /root/.flair/models/hunflair-gene-full-v1.0.pt\n",
            "2022-08-31 15:10:50,791 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Gene, B-Gene, I-Gene, E-Gene, <START>, <STOP>\n",
            "2022-08-31 15:10:51,744 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_allcorpus_models/huner-species/hunflair-species-full-v1.1.pt not found in cache, downloading to /tmp/tmpq9332g8f\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104886931/1104886931 [00:48<00:00, 22579238.44B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:11:41,074 copying /tmp/tmpq9332g8f to cache at /root/.flair/models/hunflair-species-full-v1.1.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-08-31 15:11:45,083 removing temp file /tmp/tmpq9332g8f\n",
            "2022-08-31 15:11:45,242 loading file /root/.flair/models/hunflair-species-full-v1.1.pt\n",
            "2022-08-31 15:11:52,042 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Species, B-Species, I-Species, E-Species, <START>, <STOP>\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# initialize the sentence splitter\n",
        "splitter = SciSpacySentenceSplitter()"
      ],
      "metadata": {
        "id": "rU2q0SoqC80c"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "abstracts = pd.read_csv(abstracts_file, sep='\\t', delimiter=None, names=['PMID', 'Title', 'Abstract'], dtype='str', index_col=['PMID'], encoding = \"ISO-8859-1\")"
      ],
      "metadata": {
        "id": "VCrsrfGc4zvS"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "LAYYmv8MC79E"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "abstracts.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        },
        "id": "KvweA205rWiT",
        "outputId": "94b71da4-7d7b-41f0-dc2f-2afaa899a33e"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                      Title  \\\n",
              "PMID                                                          \n",
              "31424513  Phosphoproteomic Profiling Reveals the Importa...   \n",
              "12582914  Substitution mapping of Pup1: a major QTL incr...   \n",
              "22020099  Rice GDP dissociation inhibitor 3 inhibits OsM...   \n",
              "12582916  Genetic basis of hybrid breakdown in a Japonic...   \n",
              "25821195  Quantitative trait locus analysis and fine map...   \n",
              "\n",
              "                                                   Abstract  \n",
              "PMID                                                         \n",
              "31424513  Phosphorus is one of the most important macron...  \n",
              "12582914  A major QTL for P uptake had previously been m...  \n",
              "22020099  GDP dissociation inhibitor (GDI) plays an esse...  \n",
              "12582916  Reproductive barriers often arise in hybrid pr...  \n",
              "25821195  Two QTLs were identified to control panicle le...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-05d145ed-055a-4697-8759-3eb4f3aa7841\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Title</th>\n",
              "      <th>Abstract</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PMID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>31424513</th>\n",
              "      <td>Phosphoproteomic Profiling Reveals the Importa...</td>\n",
              "      <td>Phosphorus is one of the most important macron...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12582914</th>\n",
              "      <td>Substitution mapping of Pup1: a major QTL incr...</td>\n",
              "      <td>A major QTL for P uptake had previously been m...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22020099</th>\n",
              "      <td>Rice GDP dissociation inhibitor 3 inhibits OsM...</td>\n",
              "      <td>GDP dissociation inhibitor (GDI) plays an esse...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12582916</th>\n",
              "      <td>Genetic basis of hybrid breakdown in a Japonic...</td>\n",
              "      <td>Reproductive barriers often arise in hybrid pr...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25821195</th>\n",
              "      <td>Quantitative trait locus analysis and fine map...</td>\n",
              "      <td>Two QTLs were identified to control panicle le...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-05d145ed-055a-4697-8759-3eb4f3aa7841')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-05d145ed-055a-4697-8759-3eb4f3aa7841 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-05d145ed-055a-4697-8759-3eb4f3aa7841');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "short_list = abstracts.head().copy()"
      ],
      "metadata": {
        "id": "r11RPj13qGcK"
      },
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "short_list.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        },
        "id": "NDzUT2YJvrhR",
        "outputId": "c2fd1db7-2116-4b3d-9daf-6b8be3147fc8"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                      Title  \\\n",
              "PMID                                                          \n",
              "31424513  Phosphoproteomic Profiling Reveals the Importa...   \n",
              "12582914  Substitution mapping of Pup1: a major QTL incr...   \n",
              "22020099  Rice GDP dissociation inhibitor 3 inhibits OsM...   \n",
              "12582916  Genetic basis of hybrid breakdown in a Japonic...   \n",
              "25821195  Quantitative trait locus analysis and fine map...   \n",
              "\n",
              "                                                   Abstract  \n",
              "PMID                                                         \n",
              "31424513  Phosphorus is one of the most important macron...  \n",
              "12582914  A major QTL for P uptake had previously been m...  \n",
              "22020099  GDP dissociation inhibitor (GDI) plays an esse...  \n",
              "12582916  Reproductive barriers often arise in hybrid pr...  \n",
              "25821195  Two QTLs were identified to control panicle le...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3da99a3b-ef3a-43ed-afff-01a6bc701ce0\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Title</th>\n",
              "      <th>Abstract</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PMID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>31424513</th>\n",
              "      <td>Phosphoproteomic Profiling Reveals the Importa...</td>\n",
              "      <td>Phosphorus is one of the most important macron...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12582914</th>\n",
              "      <td>Substitution mapping of Pup1: a major QTL incr...</td>\n",
              "      <td>A major QTL for P uptake had previously been m...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22020099</th>\n",
              "      <td>Rice GDP dissociation inhibitor 3 inhibits OsM...</td>\n",
              "      <td>GDP dissociation inhibitor (GDI) plays an esse...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12582916</th>\n",
              "      <td>Genetic basis of hybrid breakdown in a Japonic...</td>\n",
              "      <td>Reproductive barriers often arise in hybrid pr...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25821195</th>\n",
              "      <td>Quantitative trait locus analysis and fine map...</td>\n",
              "      <td>Two QTLs were identified to control panicle le...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3da99a3b-ef3a-43ed-afff-01a6bc701ce0')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-3da99a3b-ef3a-43ed-afff-01a6bc701ce0 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-3da99a3b-ef3a-43ed-afff-01a6bc701ce0');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#import en_core_sci_sm\n",
        "#nlp = en_core_sci_sm.load()\n",
        "for index, record in short_list.iterrows():\n",
        "  sentence = Sentence(record['Title'], use_tokenizer=SciSpacyTokenizer())\n",
        "  tagger.predict(sentence)\n",
        "  print(\"\\nPubMed: {}\\n\\nTitle: {}\\n\".format(index,record['Title']))\n",
        "  #print(sentence.to_dict())\n",
        "  for gene in sentence.get_spans(\"hunflair-gene\"):\n",
        "    #for entity in sentence.ge#t_spans(annotation_layer):\n",
        "      print(gene)\n",
        "  # split text into a list of Sentence objects\n",
        "  sentences = splitter.split(record['Abstract'])\n",
        "  # you can apply the HunFlair tagger directly to this list\n",
        "  tagger.predict(sentences)\n",
        "  for sentence in sentences:\n",
        "    print(\"\\n\")\n",
        "    #print(sentence.to_tagged_string())\n",
        "    if sentence.get_spans(\"hunflair-gene\"):\n",
        "      print(sentence.text)\n",
        "      for gene in sentence.get_spans(\"hunflair-gene\"):\n",
        "      #for entity in sentence.get_spans(annotation_layer):\n",
        "        print(gene)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OT3ir3lCrtm9",
        "outputId": "8983da7c-6f62-468a-9a07-a89ac3e08eaf"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "PubMed: 31424513\n",
            "\n",
            "Title: Phosphoproteomic Profiling Reveals the Importance of CK2, MAPKs and CDPKs in Response to Phosphate Starvation in Rice.\n",
            "\n",
            "Span[6:7]: \"CK2\" → Gene (0.9083)\n",
            "Span[8:9]: \"MAPKs\" → Gene (0.8176)\n",
            "Span[10:11]: \"CDPKs\" → Gene (0.9272)\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "Levels of phosphorylation of four mitogen - activated protein kinases ( MAPKs ) , including OsMAPK6 , five calcium - dependent protein kinases ( CDPKs ) and OsCK2Î²3 decreased in response to Pi starvation .\n",
            "Span[5:10]: \"mitogen - activated protein kinases\" → Gene (0.6933)\n",
            "Span[11:12]: \"MAPKs\" → Gene (0.9207)\n",
            "Span[15:16]: \"OsMAPK6\" → Gene (0.9508)\n",
            "Span[18:23]: \"calcium - dependent protein kinases\" → Gene (0.7019)\n",
            "Span[24:25]: \"CDPKs\" → Gene (0.9065)\n",
            "Span[27:28]: \"OsCK2Î²3\" → Gene (0.9603)\n",
            "\n",
            "\n",
            "The decreased phosphorylation level of OsMAPK6 was confirmed by Western blotting .\n",
            "Span[5:6]: \"OsMAPK6\" → Gene (0.9637)\n",
            "\n",
            "\n",
            "Mutation of OsMAPK6 led to Pi accumulation under Pi - sufficient conditions .\n",
            "Span[2:3]: \"OsMAPK6\" → Gene (0.9642)\n",
            "\n",
            "\n",
            "Motif analysis indicated that the putative MAPK , casein kinase 2 ( CK2 ) and CDPK substrates represented about 54.4 % , 21.5 % and 4.7 % , respectively , of the proteins exhibiting differential phosphorylation .\n",
            "Span[6:7]: \"MAPK\" → Gene (0.859)\n",
            "Span[8:11]: \"casein kinase 2\" → Gene (0.7095)\n",
            "Span[12:13]: \"CK2\" → Gene (0.9663)\n",
            "Span[15:16]: \"CDPK\" → Gene (0.8566)\n",
            "\n",
            "\n",
            "Based on the motif analysis , 191 , 151 and 46 candidate substrates for MAPK , CK2 and CDPK were identified .\n",
            "Span[14:15]: \"MAPK\" → Gene (0.9263)\n",
            "Span[16:17]: \"CK2\" → Gene (0.9519)\n",
            "Span[18:19]: \"CDPK\" → Gene (0.9616)\n",
            "\n",
            "\n",
            "These results indicate that modification of phosphorylation profiles provides complementary information on Pi - starvation - induced processes , with CK2 , MAPK and CDPK protein kinase families playing key roles in these processes in rice .\n",
            "Span[20:21]: \"CK2\" → Gene (0.8855)\n",
            "Span[22:23]: \"MAPK\" → Gene (0.8507)\n",
            "Span[24:28]: \"CDPK protein kinase families\" → Gene (0.622)\n",
            "\n",
            "PubMed: 12582914\n",
            "\n",
            "Title: Substitution mapping of Pup1: a major QTL increasing phosphorus uptake of rice from a phosphorus-deficient soil.\n",
            "\n",
            "Span[3:4]: \"Pup1\" → Gene (0.9566)\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "QTL mapping showed that close to 80 % of the variation between families was due to a single QTL , hereafter referred to as Pup1 ( Phosphorus uptake 1 ) .\n",
            "Span[24:25]: \"Pup1\" → Gene (0.8106)\n",
            "\n",
            "\n",
            "Pup1 was placed in a 3-cM interval flanked by markers S14025 and S13126 , which is within 1 cM of the position identified in the original QTL mapping experiment .\n",
            "Span[0:1]: \"Pup1\" → Gene (0.8562)\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "Substitution mapping revealed that Pup1 co - segregated with marker S13126 and that the flanking markers , S14025 and S13752 , were outside the interval containing Pup1 .\n",
            "Span[4:5]: \"Pup1\" → Gene (0.8521)\n",
            "Span[26:27]: \"Pup1\" → Gene (0.873)\n",
            "\n",
            "\n",
            "The two mapping strategies therefore yielded almost identical results and , in combining the advantages of both , Pup1 could be mapped with high certainty .\n",
            "Span[18:19]: \"Pup1\" → Gene (0.8392)\n",
            "\n",
            "\n",
            "\n",
            "PubMed: 22020099\n",
            "\n",
            "Title: Rice GDP dissociation inhibitor 3 inhibits OsMAPK2 activity through physical interaction.\n",
            "\n",
            "Span[0:5]: \"Rice GDP dissociation inhibitor 3\" → Gene (0.687)\n",
            "Span[6:7]: \"OsMAPK2\" → Gene (0.9456)\n",
            "\n",
            "\n",
            "GDP dissociation inhibitor ( GDI ) plays an essential role in regulating the state of bound nucleotides and subcellular localizations of Rab proteins .\n",
            "Span[0:3]: \"GDP dissociation inhibitor\" → Gene (0.7329)\n",
            "Span[4:5]: \"GDI\" → Gene (0.8254)\n",
            "Span[21:22]: \"Rab\" → Gene (0.7816)\n",
            "\n",
            "\n",
            "In our previous study , we showed that OsGDI3 facilitates the recycling of OsRab11 with a help of OsGAP1 .\n",
            "Span[8:9]: \"OsGDI3\" → Gene (0.825)\n",
            "Span[13:14]: \"OsRab11\" → Gene (0.9035)\n",
            "Span[18:19]: \"OsGAP1\" → Gene (0.924)\n",
            "\n",
            "\n",
            "In this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\n",
            "Span[7:8]: \"OsGDI3\" → Gene (0.9153)\n",
            "Span[11:14]: \"sec19 - 1\" → Gene (0.6907)\n",
            "Span[23:26]: \"yeast GDI gene\" → Gene (0.5873)\n",
            "Span[29:30]: \"OsGDI3\" → Gene (0.8895)\n",
            "Span[35:37]: \"yeast GDI\" → Gene (0.6493)\n",
            "\n",
            "\n",
            "To obtain further knowledge on the function of OsGDI3 , candidate OsGDI3-interacting proteins were identified by yeast two - hybrid screens .\n",
            "Span[8:9]: \"OsGDI3\" → Gene (0.9118)\n",
            "Span[11:12]: \"OsGDI3-interacting\" → Gene (0.7388)\n",
            "\n",
            "\n",
            "OsMAPK2 is one of OsGDI3 interacting proteins from yeast two - hybrid screens and subject to further analysis .\n",
            "Span[0:1]: \"OsMAPK2\" → Gene (0.9493)\n",
            "Span[4:5]: \"OsGDI3\" → Gene (0.8251)\n",
            "\n",
            "\n",
            "A kinase assay showed that the autophosphorylation activity of OsMAPK2 is inhibited by OsGDI3 in vitro .\n",
            "Span[1:2]: \"kinase\" → Gene (0.6095)\n",
            "Span[9:10]: \"OsMAPK2\" → Gene (0.9646)\n",
            "Span[13:14]: \"OsGDI3\" → Gene (0.7082)\n",
            "\n",
            "\n",
            "In addition , ectopic expressions of OsGDI3-in Arabidopsis cause reductions at the level of phosphorylated AtMPK in phosphorylation activity .\n",
            "Span[6:7]: \"OsGDI3-in\" → Gene (0.7466)\n",
            "Span[15:16]: \"AtMPK\" → Gene (0.8348)\n",
            "\n",
            "\n",
            "Taken together , OsGDI3 functions as a negative regulator of OsMAPK2 through modulating its kinase activity .\n",
            "Span[3:4]: \"OsGDI3\" → Gene (0.9232)\n",
            "Span[10:11]: \"OsMAPK2\" → Gene (0.9543)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "mV4Hvq10xfpC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#**Saving the results**"
      ],
      "metadata": {
        "id": "H6JTGnTm_f0H"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! git pull\n",
        "#! git add output/iric_oryzabase_tf_annotation.pkl\n",
        "! git commit -m \"saving\"\n",
        "! git push"
      ],
      "metadata": {
        "id": "V3BtDBUs_l6D",
        "outputId": "b4235cd9-4192-4e5a-8655-347fc95c0410",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "remote: Enumerating objects: 7, done.\u001b[K\n",
            "remote: Counting objects:  14% (1/7)\u001b[K\rremote: Counting objects:  28% (2/7)\u001b[K\rremote: Counting objects:  42% (3/7)\u001b[K\rremote: Counting objects:  57% (4/7)\u001b[K\rremote: Counting objects:  71% (5/7)\u001b[K\rremote: Counting objects:  85% (6/7)\u001b[K\rremote: Counting objects: 100% (7/7)\u001b[K\rremote: Counting objects: 100% (7/7), done.\u001b[K\n",
            "remote: Compressing objects:  25% (1/4)\u001b[K\rremote: Compressing objects:  50% (2/4)\u001b[K\rremote: Compressing objects:  75% (3/4)\u001b[K\rremote: Compressing objects: 100% (4/4)\u001b[K\rremote: Compressing objects: 100% (4/4), done.\u001b[K\n",
            "remote: Total 4 (delta 2), reused 0 (delta 0), pack-reused 0\u001b[K\n",
            "Unpacking objects: 100% (4/4), done.\n",
            "From https://github.com/pierrelarmande/OryzaGP\n",
            "   ff97898..996676b  main       -> origin/main\n",
            "hint: Waiting for your editor to close the file... error: unable to start editor 'editor'\n",
            "Not committing merge; use 'git commit' to complete the merge.\n",
            "[main 39228c5] saving\n",
            "fatal: could not read Password for 'https://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@github.com': No such device or address\n"
          ]
        }
      ]
    }
  ]
}