{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "OryzaGP.ipynb",
      "provenance": [],
      "authorship_tag": "ABX9TyOutmV3IgzMrYT34IuxHEvR",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pierrelarmande/OryzaGP/blob/main/src/OryzaGP.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "HhCo9j5-Y3vN"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import json\n",
        "import csv\n",
        "import pickle\n",
        "from urllib.request import urlopen\n",
        "import numpy as np\n",
        "import missingno as msno\n",
        "import nltk\n",
        "import pprint\n",
        "import os, sys, re, pprint, datetime, time\n",
        "import copy"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install scispacy\n",
        "!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.0/en_core_sci_sm-0.5.0.tar.gz"
      ],
      "metadata": {
        "id": "Pujom8sCsj0E",
        "outputId": "ac128a04-7301-408e-8c30-a5ae00fed717",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting scispacy\n",
            "  Downloading scispacy-0.5.0-py3-none-any.whl (44 kB)\n",
            "\u001b[K     |████████████████████████████████| 44 kB 2.1 MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from scispacy) (1.21.6)\n",
            "Requirement already satisfied: scikit-learn>=0.20.3 in /usr/local/lib/python3.7/dist-packages (from scispacy) (1.0.2)\n",
            "Collecting pysbd\n",
            "  Downloading pysbd-0.3.4-py3-none-any.whl (71 kB)\n",
            "\u001b[K     |████████████████████████████████| 71 kB 4.2 MB/s \n",
            "\u001b[?25hRequirement already satisfied: requests<3.0.0,>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scispacy) (2.23.0)\n",
            "Collecting nmslib>=1.7.3.6\n",
            "  Downloading nmslib-2.1.1-cp37-cp37m-manylinux2010_x86_64.whl (13.5 MB)\n",
            "\u001b[K     |████████████████████████████████| 13.5 MB 15.6 MB/s \n",
            "\u001b[?25hCollecting conllu\n",
            "  Downloading conllu-4.5.2-py2.py3-none-any.whl (16 kB)\n",
            "Collecting spacy<3.3.0,>=3.2.0\n",
            "  Downloading spacy-3.2.4-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.0 MB)\n",
            "\u001b[K     |████████████████████████████████| 6.0 MB 60.3 MB/s \n",
            "\u001b[?25hRequirement already satisfied: joblib in /usr/local/lib/python3.7/dist-packages (from scispacy) (1.1.0)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.7/dist-packages (from nmslib>=1.7.3.6->scispacy) (5.4.8)\n",
            "Collecting pybind11<2.6.2\n",
            "  Downloading pybind11-2.6.1-py2.py3-none-any.whl (188 kB)\n",
            "\u001b[K     |████████████████████████████████| 188 kB 43.3 MB/s \n",
            "\u001b[?25hRequirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (2022.6.15)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (2.10)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (1.24.3)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.0.0->scispacy) (3.0.4)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.20.3->scispacy) (3.1.0)\n",
            "Requirement already satisfied: scipy>=1.1.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.20.3->scispacy) (1.7.3)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (2.0.6)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (3.0.7)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (21.3)\n",
            "Requirement already satisfied: pathy>=0.3.5 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (0.6.2)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (57.4.0)\n",
            "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.8 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (3.0.10)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (4.64.0)\n",
            "Collecting pydantic!=1.8,!=1.8.1,<1.9.0,>=1.7.4\n",
            "  Downloading pydantic-1.8.2-cp37-cp37m-manylinux2014_x86_64.whl (10.1 MB)\n",
            "\u001b[K     |████████████████████████████████| 10.1 MB 53.9 MB/s \n",
            "\u001b[?25hRequirement already satisfied: srsly<3.0.0,>=2.4.1 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (2.4.4)\n",
            "Requirement already satisfied: wasabi<1.1.0,>=0.8.1 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (0.10.1)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (1.0.8)\n",
            "Requirement already satisfied: typer<0.5.0,>=0.3.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (0.4.2)\n",
            "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (2.0.8)\n",
            "Requirement already satisfied: blis<0.8.0,>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (0.7.8)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (2.11.3)\n",
            "Requirement already satisfied: langcodes<4.0.0,>=3.2.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (3.3.0)\n",
            "Requirement already satisfied: click<8.1.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (7.1.2)\n",
            "Collecting thinc<8.1.0,>=8.0.12\n",
            "  Downloading thinc-8.0.17-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (660 kB)\n",
            "\u001b[K     |████████████████████████████████| 660 kB 9.4 MB/s \n",
            "\u001b[?25hCollecting typing-extensions<4.0.0.0,>=3.7.4\n",
            "  Downloading typing_extensions-3.10.0.2-py3-none-any.whl (26 kB)\n",
            "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.0->scispacy) (1.0.3)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from catalogue<2.1.0,>=2.0.6->spacy<3.3.0,>=3.2.0->scispacy) (3.8.1)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging>=20.0->spacy<3.3.0,>=3.2.0->scispacy) (3.0.9)\n",
            "Requirement already satisfied: smart-open<6.0.0,>=5.2.1 in /usr/local/lib/python3.7/dist-packages (from pathy>=0.3.5->spacy<3.3.0,>=3.2.0->scispacy) (5.2.1)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.7/dist-packages (from jinja2->spacy<3.3.0,>=3.2.0->scispacy) (2.0.1)\n",
            "Installing collected packages: typing-extensions, pydantic, thinc, pybind11, spacy, pysbd, nmslib, conllu, scispacy\n",
            "  Attempting uninstall: typing-extensions\n",
            "    Found existing installation: typing-extensions 4.1.1\n",
            "    Uninstalling typing-extensions-4.1.1:\n",
            "      Successfully uninstalled typing-extensions-4.1.1\n",
            "  Attempting uninstall: pydantic\n",
            "    Found existing installation: pydantic 1.9.2\n",
            "    Uninstalling pydantic-1.9.2:\n",
            "      Successfully uninstalled pydantic-1.9.2\n",
            "  Attempting uninstall: thinc\n",
            "    Found existing installation: thinc 8.1.0\n",
            "    Uninstalling thinc-8.1.0:\n",
            "      Successfully uninstalled thinc-8.1.0\n",
            "  Attempting uninstall: spacy\n",
            "    Found existing installation: spacy 3.4.1\n",
            "    Uninstalling spacy-3.4.1:\n",
            "      Successfully uninstalled spacy-3.4.1\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "en-core-web-sm 3.4.0 requires spacy<3.5.0,>=3.4.0, but you have spacy 3.2.4 which is incompatible.\u001b[0m\n",
            "Successfully installed conllu-4.5.2 nmslib-2.1.1 pybind11-2.6.1 pydantic-1.8.2 pysbd-0.3.4 scispacy-0.5.0 spacy-3.2.4 thinc-8.0.17 typing-extensions-3.10.0.2\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "typing_extensions"
                ]
              }
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.0/en_core_sci_sm-0.5.0.tar.gz\n",
            "  Downloading https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.0/en_core_sci_sm-0.5.0.tar.gz (15.9 MB)\n",
            "\u001b[K     |████████████████████████████████| 15.9 MB 3.5 MB/s \n",
            "\u001b[?25hRequirement already satisfied: spacy<3.3.0,>=3.2.3 in /usr/local/lib/python3.7/dist-packages (from en-core-sci-sm==0.5.0) (3.2.4)\n",
            "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (4.64.0)\n",
            "Requirement already satisfied: typing-extensions<4.0.0.0,>=3.7.4 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.10.0.2)\n",
            "Requirement already satisfied: typer<0.5.0,>=0.3.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (0.4.2)\n",
            "Requirement already satisfied: thinc<8.1.0,>=8.0.12 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (8.0.17)\n",
            "Requirement already satisfied: click<8.1.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (7.1.2)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (21.3)\n",
            "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.0.7)\n",
            "Requirement already satisfied: pathy>=0.3.5 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (0.6.2)\n",
            "Requirement already satisfied: pydantic!=1.8,!=1.8.1,<1.9.0,>=1.7.4 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.8.2)\n",
            "Requirement already satisfied: spacy-legacy<3.1.0,>=3.0.8 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.0.10)\n",
            "Requirement already satisfied: wasabi<1.1.0,>=0.8.1 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (0.10.1)\n",
            "Requirement already satisfied: requests<3.0.0,>=2.13.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.23.0)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (57.4.0)\n",
            "Requirement already satisfied: spacy-loggers<2.0.0,>=1.0.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.0.3)\n",
            "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.0.8)\n",
            "Requirement already satisfied: numpy>=1.15.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.21.6)\n",
            "Requirement already satisfied: srsly<3.0.0,>=2.4.1 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.4.4)\n",
            "Requirement already satisfied: blis<0.8.0,>=0.4.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (0.7.8)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.11.3)\n",
            "Requirement already satisfied: langcodes<4.0.0,>=3.2.0 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.3.0)\n",
            "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.0.6)\n",
            "Requirement already satisfied: catalogue<2.1.0,>=2.0.6 in /usr/local/lib/python3.7/dist-packages (from spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.0.8)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from catalogue<2.1.0,>=2.0.6->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.8.1)\n",
            "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /usr/local/lib/python3.7/dist-packages (from packaging>=20.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.0.9)\n",
            "Requirement already satisfied: smart-open<6.0.0,>=5.2.1 in /usr/local/lib/python3.7/dist-packages (from pathy>=0.3.5->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (5.2.1)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2022.6.15)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (3.0.4)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (1.24.3)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests<3.0.0,>=2.13.0->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.10)\n",
            "Requirement already satisfied: MarkupSafe>=0.23 in /usr/local/lib/python3.7/dist-packages (from jinja2->spacy<3.3.0,>=3.2.3->en-core-sci-sm==0.5.0) (2.0.1)\n",
            "Building wheels for collected packages: en-core-sci-sm\n",
            "  Building wheel for en-core-sci-sm (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for en-core-sci-sm: filename=en_core_sci_sm-0.5.0-py3-none-any.whl size=15871046 sha256=1014f43f477dd7613e1f3cd03c1f47d5012ee386b5fb71873d1a7329a4b9d860\n",
            "  Stored in directory: /root/.cache/pip/wheels/b1/b8/7d/d26a5894db99bf0cfb9add399889738de91a608e6664f82454\n",
            "Successfully built en-core-sci-sm\n",
            "Installing collected packages: en-core-sci-sm\n",
            "Successfully installed en-core-sci-sm-0.5.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install flair"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LRn7z0Osr7Rm",
        "outputId": "4985179e-9f94-40a3-8c90-2b9737365b01"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting flair\n",
            "  Downloading flair-0.11.3-py3-none-any.whl (401 kB)\n",
            "\u001b[K     |████████████████████████████████| 401 kB 8.8 MB/s \n",
            "\u001b[?25hCollecting sqlitedict>=1.6.0\n",
            "  Downloading sqlitedict-2.0.0.tar.gz (46 kB)\n",
            "\u001b[K     |████████████████████████████████| 46 kB 4.1 MB/s \n",
            "\u001b[?25hCollecting mpld3==0.3\n",
            "  Downloading mpld3-0.3.tar.gz (788 kB)\n",
            "\u001b[K     |████████████████████████████████| 788 kB 52.2 MB/s \n",
            "\u001b[?25hCollecting janome\n",
            "  Downloading Janome-0.4.2-py2.py3-none-any.whl (19.7 MB)\n",
            "\u001b[K     |████████████████████████████████| 19.7 MB 5.2 MB/s \n",
            "\u001b[?25hCollecting sentencepiece==0.1.95\n",
            "  Downloading sentencepiece-0.1.95-cp37-cp37m-manylinux2014_x86_64.whl (1.2 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.2 MB 51.4 MB/s \n",
            "\u001b[?25hRequirement already satisfied: more-itertools in /usr/local/lib/python3.7/dist-packages (from flair) (8.14.0)\n",
            "Requirement already satisfied: python-dateutil>=2.6.1 in /usr/local/lib/python3.7/dist-packages (from flair) (2.8.2)\n",
            "Collecting ftfy\n",
            "  Downloading ftfy-6.1.1-py3-none-any.whl (53 kB)\n",
            "\u001b[K     |████████████████████████████████| 53 kB 1.9 MB/s \n",
            "\u001b[?25hCollecting deprecated>=1.2.4\n",
            "  Downloading Deprecated-1.2.13-py2.py3-none-any.whl (9.6 kB)\n",
            "Requirement already satisfied: regex in /usr/local/lib/python3.7/dist-packages (from flair) (2022.6.2)\n",
            "Collecting wikipedia-api\n",
            "  Downloading Wikipedia-API-0.5.4.tar.gz (18 kB)\n",
            "Requirement already satisfied: gensim>=3.4.0 in /usr/local/lib/python3.7/dist-packages (from flair) (3.6.0)\n",
            "Requirement already satisfied: lxml in /usr/local/lib/python3.7/dist-packages (from flair) (4.9.1)\n",
            "Collecting hyperopt>=0.2.7\n",
            "  Downloading hyperopt-0.2.7-py2.py3-none-any.whl (1.6 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.6 MB 42.2 MB/s \n",
            "\u001b[?25hCollecting huggingface-hub\n",
            "  Downloading huggingface_hub-0.9.1-py3-none-any.whl (120 kB)\n",
            "\u001b[K     |████████████████████████████████| 120 kB 51.3 MB/s \n",
            "\u001b[?25hRequirement already satisfied: matplotlib>=2.2.3 in /usr/local/lib/python3.7/dist-packages (from flair) (3.2.2)\n",
            "Collecting transformers>=4.0.0\n",
            "  Downloading transformers-4.21.2-py3-none-any.whl (4.7 MB)\n",
            "\u001b[K     |████████████████████████████████| 4.7 MB 38.3 MB/s \n",
            "\u001b[?25hCollecting bpemb>=0.3.2\n",
            "  Downloading bpemb-0.3.3-py3-none-any.whl (19 kB)\n",
            "Requirement already satisfied: conllu>=4.0 in /usr/local/lib/python3.7/dist-packages (from flair) (4.5.2)\n",
            "Collecting konoha<5.0.0,>=4.0.0\n",
            "  Downloading konoha-4.6.5-py3-none-any.whl (20 kB)\n",
            "Collecting pptree\n",
            "  Downloading pptree-3.1.tar.gz (3.0 kB)\n",
            "Requirement already satisfied: scikit-learn>=0.21.3 in /usr/local/lib/python3.7/dist-packages (from flair) (1.0.2)\n",
            "Requirement already satisfied: torch!=1.8,>=1.5.0 in /usr/local/lib/python3.7/dist-packages (from flair) (1.12.1+cu113)\n",
            "Requirement already satisfied: gdown==4.4.0 in /usr/local/lib/python3.7/dist-packages (from flair) (4.4.0)\n",
            "Requirement already satisfied: tqdm>=4.26.0 in /usr/local/lib/python3.7/dist-packages (from flair) (4.64.0)\n",
            "Collecting segtok>=1.5.7\n",
            "  Downloading segtok-1.5.11-py3-none-any.whl (24 kB)\n",
            "Requirement already satisfied: tabulate in /usr/local/lib/python3.7/dist-packages (from flair) (0.8.10)\n",
            "Collecting langdetect\n",
            "  Downloading langdetect-1.0.9.tar.gz (981 kB)\n",
            "\u001b[K     |████████████████████████████████| 981 kB 40.0 MB/s \n",
            "\u001b[?25hRequirement already satisfied: beautifulsoup4 in /usr/local/lib/python3.7/dist-packages (from gdown==4.4.0->flair) (4.6.3)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.7/dist-packages (from gdown==4.4.0->flair) (3.8.0)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from gdown==4.4.0->flair) (1.15.0)\n",
            "Requirement already satisfied: requests[socks] in /usr/local/lib/python3.7/dist-packages (from gdown==4.4.0->flair) (2.23.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from bpemb>=0.3.2->flair) (1.21.6)\n",
            "Requirement already satisfied: wrapt<2,>=1.10 in /usr/local/lib/python3.7/dist-packages (from deprecated>=1.2.4->flair) (1.14.1)\n",
            "Requirement already satisfied: scipy>=0.18.1 in /usr/local/lib/python3.7/dist-packages (from gensim>=3.4.0->flair) (1.7.3)\n",
            "Requirement already satisfied: smart-open>=1.2.1 in /usr/local/lib/python3.7/dist-packages (from gensim>=3.4.0->flair) (5.2.1)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.7/dist-packages (from hyperopt>=0.2.7->flair) (0.16.0)\n",
            "Requirement already satisfied: cloudpickle in /usr/local/lib/python3.7/dist-packages (from hyperopt>=0.2.7->flair) (1.5.0)\n",
            "Requirement already satisfied: networkx>=2.2 in /usr/local/lib/python3.7/dist-packages (from hyperopt>=0.2.7->flair) (2.6.3)\n",
            "Collecting py4j\n",
            "  Downloading py4j-0.10.9.7-py2.py3-none-any.whl (200 kB)\n",
            "\u001b[K     |████████████████████████████████| 200 kB 52.2 MB/s \n",
            "\u001b[?25hCollecting requests\n",
            "  Downloading requests-2.28.1-py3-none-any.whl (62 kB)\n",
            "\u001b[K     |████████████████████████████████| 62 kB 1.6 MB/s \n",
            "\u001b[?25hCollecting importlib-metadata<4.0.0,>=3.7.0\n",
            "  Downloading importlib_metadata-3.10.1-py3-none-any.whl (14 kB)\n",
            "Collecting overrides<4.0.0,>=3.0.0\n",
            "  Downloading overrides-3.1.0.tar.gz (11 kB)\n",
            "Requirement already satisfied: typing-extensions>=3.6.4 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata<4.0.0,>=3.7.0->konoha<5.0.0,>=4.0.0->flair) (3.10.0.2)\n",
            "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/dist-packages (from importlib-metadata<4.0.0,>=3.7.0->konoha<5.0.0,>=4.0.0->flair) (3.8.1)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=2.2.3->flair) (1.4.4)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=2.2.3->flair) (3.0.9)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=2.2.3->flair) (0.11.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (2022.6.15)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (1.24.3)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (2.10)\n",
            "Requirement already satisfied: charset-normalizer<3,>=2 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (2.1.1)\n",
            "Requirement already satisfied: joblib>=0.11 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.21.3->flair) (1.1.0)\n",
            "Requirement already satisfied: threadpoolctl>=2.0.0 in /usr/local/lib/python3.7/dist-packages (from scikit-learn>=0.21.3->flair) (3.1.0)\n",
            "Collecting tokenizers!=0.11.3,<0.13,>=0.11.1\n",
            "  Downloading tokenizers-0.12.1-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (6.6 MB)\n",
            "\u001b[K     |████████████████████████████████| 6.6 MB 35.3 MB/s \n",
            "\u001b[?25hRequirement already satisfied: pyyaml>=5.1 in /usr/local/lib/python3.7/dist-packages (from transformers>=4.0.0->flair) (6.0)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.7/dist-packages (from transformers>=4.0.0->flair) (21.3)\n",
            "Requirement already satisfied: wcwidth>=0.2.5 in /usr/local/lib/python3.7/dist-packages (from ftfy->flair) (0.2.5)\n",
            "Requirement already satisfied: PySocks!=1.5.7,>=1.5.6 in /usr/local/lib/python3.7/dist-packages (from requests->bpemb>=0.3.2->flair) (1.7.1)\n",
            "Building wheels for collected packages: mpld3, overrides, sqlitedict, langdetect, pptree, wikipedia-api\n",
            "  Building wheel for mpld3 (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for mpld3: filename=mpld3-0.3-py3-none-any.whl size=116702 sha256=ebb0fea691ded8eb9d753048c8e616d5c16b7d308edef8bff90509b0a02997ac\n",
            "  Stored in directory: /root/.cache/pip/wheels/26/70/6a/1c79e59951a41b4045497da187b2724f5659ca64033cf4548e\n",
            "  Building wheel for overrides (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for overrides: filename=overrides-3.1.0-py3-none-any.whl size=10187 sha256=6ef5f3e9ca1158f71a74a5bd9f047ede4decee62e0b29d18e3c7af69276efff5\n",
            "  Stored in directory: /root/.cache/pip/wheels/3a/0d/38/01a9bc6e20dcfaf0a6a7b552d03137558ba1c38aea47644682\n",
            "  Building wheel for sqlitedict (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for sqlitedict: filename=sqlitedict-2.0.0-py3-none-any.whl size=15736 sha256=64b1a00d628630a252d1422d855d96676de642dab83081b0fa9c3e5276173eed\n",
            "  Stored in directory: /root/.cache/pip/wheels/96/dd/2e/0ed4a25cb73fc30c7ea8d10b50acb7226175736067e40a7ea3\n",
            "  Building wheel for langdetect (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for langdetect: filename=langdetect-1.0.9-py3-none-any.whl size=993242 sha256=ca70d7e514bd4ade304a8c85bf6b317f09d881d5f0a99b1a73379b8d589378c1\n",
            "  Stored in directory: /root/.cache/pip/wheels/c5/96/8a/f90c59ed25d75e50a8c10a1b1c2d4c402e4dacfa87f3aff36a\n",
            "  Building wheel for pptree (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pptree: filename=pptree-3.1-py3-none-any.whl size=4629 sha256=20889949604789f03da850d89f273acfa229bdbcc28ca237eb728e0c71cac9dc\n",
            "  Stored in directory: /root/.cache/pip/wheels/9e/e8/7d/a9c3c19b4722608a0d8b05a38c36bc3f230c43becd2a46794b\n",
            "  Building wheel for wikipedia-api (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for wikipedia-api: filename=Wikipedia_API-0.5.4-py3-none-any.whl size=13477 sha256=1ed874cbc0b2e9762140061799a916ced5705842c285d0a8c20d2941a7a63cd9\n",
            "  Stored in directory: /root/.cache/pip/wheels/d3/24/56/58ba93cf78be162451144e7a9889603f437976ef1ae7013d04\n",
            "Successfully built mpld3 overrides sqlitedict langdetect pptree wikipedia-api\n",
            "Installing collected packages: requests, importlib-metadata, tokenizers, sentencepiece, py4j, overrides, huggingface-hub, wikipedia-api, transformers, sqlitedict, segtok, pptree, mpld3, langdetect, konoha, janome, hyperopt, ftfy, deprecated, bpemb, flair\n",
            "  Attempting uninstall: requests\n",
            "    Found existing installation: requests 2.23.0\n",
            "    Uninstalling requests-2.23.0:\n",
            "      Successfully uninstalled requests-2.23.0\n",
            "  Attempting uninstall: importlib-metadata\n",
            "    Found existing installation: importlib-metadata 4.12.0\n",
            "    Uninstalling importlib-metadata-4.12.0:\n",
            "      Successfully uninstalled importlib-metadata-4.12.0\n",
            "  Attempting uninstall: hyperopt\n",
            "    Found existing installation: hyperopt 0.1.2\n",
            "    Uninstalling hyperopt-0.1.2:\n",
            "      Successfully uninstalled hyperopt-0.1.2\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "markdown 3.4.1 requires importlib-metadata>=4.4; python_version < \"3.10\", but you have importlib-metadata 3.10.1 which is incompatible.\n",
            "gym 0.25.2 requires importlib-metadata>=4.8.0; python_version < \"3.10\", but you have importlib-metadata 3.10.1 which is incompatible.\n",
            "en-core-web-sm 3.4.0 requires spacy<3.5.0,>=3.4.0, but you have spacy 3.2.4 which is incompatible.\u001b[0m\n",
            "Successfully installed bpemb-0.3.3 deprecated-1.2.13 flair-0.11.3 ftfy-6.1.1 huggingface-hub-0.9.1 hyperopt-0.2.7 importlib-metadata-3.10.1 janome-0.4.2 konoha-4.6.5 langdetect-1.0.9 mpld3-0.3 overrides-3.1.0 pptree-3.1 py4j-0.10.9.7 requests-2.28.1 segtok-1.5.11 sentencepiece-0.1.95 sqlitedict-2.0.0 tokenizers-0.12.1 transformers-4.21.2 wikipedia-api-0.5.4\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "2gtZwz-ksFpO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import requests\n",
        "from bs4 import BeautifulSoup\n",
        "#from pubmed_lookup import PubMedLookup\n",
        "#from pubmed_lookup import Publication\n",
        "from joblib import Parallel, delayed"
      ],
      "metadata": {
        "id": "PZ6Bfiyxsluz"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')\n",
        "%cd /content/gdrive/MyDrive/Github/"
      ],
      "metadata": {
        "id": "zC5l8iHFd9uT",
        "outputId": "8a30e60b-c389-4425-b9e1-20cae7341c8c",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/gdrive\n",
            "/content/gdrive/MyDrive/Github\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Connecting to the GIT repository\n",
        "! git config --global user.email \"plarmande@gmail.com\"\n",
        "! git config --global user.name \"pierrelarmande\"\n",
        "\n",
        "username = 'pierrelarmande'\n",
        "repo = 'OryzaGP'\n",
        "# To create a token : https://github.com/settings/tokens\n",
        "git_token = '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'\n",
        "\n",
        "url = \"https://\"+git_token+\"@github.com/\"+username+\"/\"+repo+\".git\""
      ],
      "metadata": {
        "id": "MIH7MBMLeDz3"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Clone if doesn't already exist\n",
        "#!git clone {url}\n",
        "%pwd\n",
        "%cd $repo\n",
        "#%cd $repo\n",
        "#! git pull"
      ],
      "metadata": {
        "id": "rjPAm2MOfm9T",
        "outputId": "ac0e09e5-0e4e-4601-dc37-d1dc170492cc",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/gdrive/MyDrive/Github/OryzaGP\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Get data\n",
        "reference_file = r'https://shigen.nig.ac.jp/rice/oryzabase/reference/download'\n",
        "gene_file = r'https://shigen.nig.ac.jp/rice/oryzabase/gene/download?classtag=GENE_EN_LIST'\n",
        "reference_file = 'https://shigen.nig.ac.jp/rice/oryzabase/reference/download'\n",
        "base_url = \"https://shigen.nig.ac.jp/rice/oryzabase/gene/detail/\"\n",
        "#page = requests.get(base_url + \"477\")\n",
        "output_file  = 'reference_cleaned.txt'\n",
        "pub_mappings_geneid = 'pub_mappings_genid.txt'\n",
        "abstracts_file = 'output/abstracts.tsv'\n",
        "\n",
        "#iric = pd.DataFrame(pd.read_pickle('output/iric_oryzabase_tf_rapdb_annotation.pkl'))\n",
        "!ls output"
      ],
      "metadata": {
        "id": "NA3H-miMftVJ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ff9165e6-550a-4649-b054-d9cfec6ccc35"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "abstracts.csv  abstracts.tsv  abstracts.xml  info.md\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pd_gene = pd.read_csv(gene_file, sep='\\t', delimiter=None, index_col= 0, dtype='str', encoding = \"ISO-8859-1\", error_bad_lines=False, skip_blank_lines=True)"
      ],
      "metadata": {
        "id": "DGWcbrPC6lVr",
        "outputId": "0adec16f-19bd-4d52-8ea6-e2ea0aef327a",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.\n",
            "\n",
            "\n",
            "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "o_lr8Kpw7Vly"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "array = pd.read_csv(reference_file, sep='\\t', delimiter=None, index_col= 0, dtype='str', encoding = \"ISO-8859-1\", error_bad_lines=False, skip_blank_lines=True)"
      ],
      "metadata": {
        "id": "DfqgIBTWlt97",
        "outputId": "4595fcb8-097d-4b53-d1f0-51e64f407306",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.\n",
            "\n",
            "\n",
            "  exec(code_obj, self.user_global_ns, self.user_ns)\n",
            "b'Skipping line 11636: expected 10 fields, saw 12\\n'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "array.head()"
      ],
      "metadata": {
        "id": "iovO4pnklx8e",
        "outputId": "598a3460-85db-432f-aa20-1a77cf778241",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 368
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "             PubMedId                                             Author  \\\n",
              "Reference_id                                                               \n",
              "1                 NaN  Abenes, M.L.P., R.E. Tabien, S.R. McCouch, R. ...   \n",
              "2                 NaN                                         Adachi, T.   \n",
              "3             8103599                          Ahn, S. and S.D. Tanksley   \n",
              "4                 NaN  Ahn, S.N., C.N. Bollich, A.M. McClung and S.D....   \n",
              "5             7903411  Ahn, S., J.A. Anderson, N.E. Sorrels and S.D. ...   \n",
              "\n",
              "                                                          Title  \\\n",
              "Reference_id                                                      \n",
              "1             Orientation and integration of the classical a...   \n",
              "2                 Dwarfness and leafy head newly found in rice.   \n",
              "3             Comparative linkage maps of the rice and maize...   \n",
              "4             RFLP analysis of genomic regions associated wi...   \n",
              "5             Homologous relationships of rice, wheat and ma...   \n",
              "\n",
              "                                    Journal Volume       Pages  Year  \\\n",
              "Reference_id                                                           \n",
              "1                                 Euphytica     76      81-87.  1994   \n",
              "2                            Agric. & Hort.  10(4)  1048-1050.  1935   \n",
              "3             Proc. Natl. Acad. Sci. U.S.A.     90  7980-7984.  1993   \n",
              "4                       Theor. Appl. Genet.     87      27-32.  1993   \n",
              "5                          Mol. Gen. Genet.    241    483-490.  1993   \n",
              "\n",
              "                CGSNL Gene Symbol  \\\n",
              "Reference_id                        \n",
              "1             LA,D27,Z2,XA21,ADH1   \n",
              "2                             LHD   \n",
              "3                             NaN   \n",
              "4                SK2(T)(SCL, FGR)   \n",
              "5                             NaN   \n",
              "\n",
              "                                              Gene Name Synonym  \n",
              "Reference_id                                                     \n",
              "1             la,d27(d t), dt, dwf23, d27, OsD27,z2,Xa21*, X...  \n",
              "2                                                           lhd  \n",
              "3                                                           NaN  \n",
              "4             sk2(t)(scl, fgr), fgr, scl, BAD2, BADH2, sk2, ...  \n",
              "5                                                           NaN  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-a8115f11-599f-48fa-aca7-ce2384517a80\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>PubMedId</th>\n",
              "      <th>Author</th>\n",
              "      <th>Title</th>\n",
              "      <th>Journal</th>\n",
              "      <th>Volume</th>\n",
              "      <th>Pages</th>\n",
              "      <th>Year</th>\n",
              "      <th>CGSNL Gene Symbol</th>\n",
              "      <th>Gene Name Synonym</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Reference_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Abenes, M.L.P., R.E. Tabien, S.R. McCouch, R. ...</td>\n",
              "      <td>Orientation and integration of the classical a...</td>\n",
              "      <td>Euphytica</td>\n",
              "      <td>76</td>\n",
              "      <td>81-87.</td>\n",
              "      <td>1994</td>\n",
              "      <td>LA,D27,Z2,XA21,ADH1</td>\n",
              "      <td>la,d27(d t), dt, dwf23, d27, OsD27,z2,Xa21*, X...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Adachi, T.</td>\n",
              "      <td>Dwarfness and leafy head newly found in rice.</td>\n",
              "      <td>Agric. &amp; Hort.</td>\n",
              "      <td>10(4)</td>\n",
              "      <td>1048-1050.</td>\n",
              "      <td>1935</td>\n",
              "      <td>LHD</td>\n",
              "      <td>lhd</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>8103599</td>\n",
              "      <td>Ahn, S. and S.D. Tanksley</td>\n",
              "      <td>Comparative linkage maps of the rice and maize...</td>\n",
              "      <td>Proc. Natl. Acad. Sci. U.S.A.</td>\n",
              "      <td>90</td>\n",
              "      <td>7980-7984.</td>\n",
              "      <td>1993</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>NaN</td>\n",
              "      <td>Ahn, S.N., C.N. Bollich, A.M. McClung and S.D....</td>\n",
              "      <td>RFLP analysis of genomic regions associated wi...</td>\n",
              "      <td>Theor. Appl. Genet.</td>\n",
              "      <td>87</td>\n",
              "      <td>27-32.</td>\n",
              "      <td>1993</td>\n",
              "      <td>SK2(T)(SCL, FGR)</td>\n",
              "      <td>sk2(t)(scl, fgr), fgr, scl, BAD2, BADH2, sk2, ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>7903411</td>\n",
              "      <td>Ahn, S., J.A. Anderson, N.E. Sorrels and S.D. ...</td>\n",
              "      <td>Homologous relationships of rice, wheat and ma...</td>\n",
              "      <td>Mol. Gen. Genet.</td>\n",
              "      <td>241</td>\n",
              "      <td>483-490.</td>\n",
              "      <td>1993</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-a8115f11-599f-48fa-aca7-ce2384517a80')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-a8115f11-599f-48fa-aca7-ce2384517a80 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-a8115f11-599f-48fa-aca7-ce2384517a80');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "array.count()"
      ],
      "metadata": {
        "id": "-lAoncTFl6U7",
        "outputId": "f88a6433-f573-4169-ab1f-365088aa9880",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PubMedId             19898\n",
              "Author               21936\n",
              "Title                21944\n",
              "Journal              21930\n",
              "Volume               20053\n",
              "Pages                19300\n",
              "Year                 21913\n",
              "CGSNL Gene Symbol     8645\n",
              "Gene Name Synonym     8577\n",
              "dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#**Clean up the Reference dataset**"
      ],
      "metadata": {
        "id": "NYNvUcStmJfd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "result = array.dropna(subset=['PubMedId'])"
      ],
      "metadata": {
        "id": "p0PIqTjamd8b"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "result.count()"
      ],
      "metadata": {
        "id": "b-7mBLfhmiQg",
        "outputId": "49927ffd-ccb5-463e-aba3-97d9131e6d20",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PubMedId             19898\n",
              "Author               19879\n",
              "Title                19886\n",
              "Journal              19872\n",
              "Volume               18153\n",
              "Pages                17252\n",
              "Year                 19855\n",
              "CGSNL Gene Symbol     7174\n",
              "Gene Name Synonym     7125\n",
              "dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# donnees.drop(['Unnamed: 15', 'Unnamed: 16'], axis=1,inplace=True)\n",
        "final_result = result.drop(['Author','Title','Journal','Volume','Pages','Year'], axis=1)"
      ],
      "metadata": {
        "id": "1INmsLfamr5L"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "final_result.head()"
      ],
      "metadata": {
        "id": "BfYUHeRzm_yD",
        "outputId": "e7b6753e-86fd-4248-fbff-8addf4751afe",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "              PubMedId CGSNL Gene Symbol  \\\n",
              "Reference_id                               \n",
              "3              8103599               NaN   \n",
              "5              7903411               NaN   \n",
              "42             7247938              ACP1   \n",
              "45            17248812           SD2,SD3   \n",
              "113           18470071          EF,BPH10   \n",
              "\n",
              "                                    Gene Name Synonym  \n",
              "Reference_id                                           \n",
              "3                                                 NaN  \n",
              "5                                                 NaN  \n",
              "42                                               Acp1  \n",
              "45                                            sd2,sd3  \n",
              "113           Ef(t)*, ef(t)*, Ef, ef,Bph10(t)*, Bph10  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-feaf9a56-ca44-4a7f-97df-ff9f5edc707f\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>PubMedId</th>\n",
              "      <th>CGSNL Gene Symbol</th>\n",
              "      <th>Gene Name Synonym</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Reference_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>8103599</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>7903411</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>7247938</td>\n",
              "      <td>ACP1</td>\n",
              "      <td>Acp1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>17248812</td>\n",
              "      <td>SD2,SD3</td>\n",
              "      <td>sd2,sd3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>113</th>\n",
              "      <td>18470071</td>\n",
              "      <td>EF,BPH10</td>\n",
              "      <td>Ef(t)*, ef(t)*, Ef, ef,Bph10(t)*, Bph10</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-feaf9a56-ca44-4a7f-97df-ff9f5edc707f')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-feaf9a56-ca44-4a7f-97df-ff9f5edc707f button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-feaf9a56-ca44-4a7f-97df-ff9f5edc707f');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final_result=final_result.fillna('_')"
      ],
      "metadata": {
        "id": "AUF1lH7qmt9L"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(len(final_result[final_result['CGSNL Gene Symbol'] == '_']))"
      ],
      "metadata": {
        "id": "K33Ah_0Smzvn",
        "outputId": "f5263435-34f0-41a3-d508-5047c7577c78",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "13081\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final_result.astype({'PubMedId': 'int32'}).dtypes"
      ],
      "metadata": {
        "id": "eg-3qbTLT4wL",
        "outputId": "b7af2e40-a523-47f7-eae4-61cb7e34c3bb",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PubMedId              int32\n",
              "CGSNL Gene Symbol    object\n",
              "Gene Name Synonym    object\n",
              "dtype: object"
            ]
          },
          "metadata": {},
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final_result['PubMedId'] = pd.to_numeric(final_result['PubMedId'])"
      ],
      "metadata": {
        "id": "EEsPhh0hWmyO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "final_result['PubMedId'].value_counts()[0]"
      ],
      "metadata": {
        "id": "3TRIpbbZUecB",
        "outputId": "42a186e3-a355-408d-8c88-4fd1b9ba3cd4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "2116"
            ]
          },
          "metadata": {},
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pmids_list = final_result['PubMedId'].to_numpy()"
      ],
      "metadata": {
        "id": "Ps7M1dGh-rd1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "final_result.dtypes"
      ],
      "metadata": {
        "id": "TAgu0kQIV1oD",
        "outputId": "3c4f447c-d0e5-49d0-8411-672efda0a4cf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PubMedId              int64\n",
              "CGSNL Gene Symbol    object\n",
              "Gene Name Synonym    object\n",
              "dtype: object"
            ]
          },
          "metadata": {},
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pmids_list"
      ],
      "metadata": {
        "id": "k6uovoncS_W-",
        "outputId": "300cfc69-fe4b-4261-88c2-c7f57ec17f93",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([ 8103599,  7903411,  7247938, ..., 35642508, 35812946, 35832225])"
            ]
          },
          "metadata": {},
          "execution_count": 23
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pmids_set = set(pmids_list)"
      ],
      "metadata": {
        "id": "b5HXzTfQXLGz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pmids = list(pmids_set)"
      ],
      "metadata": {
        "id": "NwCUHb85XOad"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pmids[1:10]"
      ],
      "metadata": {
        "id": "psSJ5n1S1H0a",
        "outputId": "1f916358-6ff6-463e-b028-20b93fa7a069",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "[31424513,\n",
              " 12582914,\n",
              " 22020099,\n",
              " 12582916,\n",
              " 25821195,\n",
              " 20938764,\n",
              " 25755661,\n",
              " 18612237,\n",
              " 31621135]"
            ]
          },
          "metadata": {},
          "execution_count": 26
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Exporting mapping Gene_ID and Gene Name"
      ],
      "metadata": {
        "id": "VgU8tfUK8E1V"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "pd_gene.head(3)"
      ],
      "metadata": {
        "id": "x2gqoBbk8NDY",
        "outputId": "1baa38c8-7dee-48d0-b2bb-943ee6a059a6",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 171
        }
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-17-7bc19fbd8e1d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpd_gene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m: name 'pd_gene' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "pd_gene.columns = [str(x).replace('(','_').replace(\n",
        "                     ' ','_').replace(\n",
        "                     '.','').replace(\n",
        "                    ')','') for x in list(pd_gene.axes[1])]"
      ],
      "metadata": {
        "id": "lnqGisv58XD1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pd_gene.columns"
      ],
      "metadata": {
        "id": "FL22y0XG8kSC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pd_mapping = pd_gene[['CGSNL_Gene_Name', 'Gene_name_synonym_s','RAP_ID', 'MSU_ID']]"
      ],
      "metadata": {
        "id": "uVDY82e68wCO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pd_mapping = pd_mapping.reset_index()"
      ],
      "metadata": {
        "id": "kbyHvrLz9HD0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pd_mapping.head()"
      ],
      "metadata": {
        "id": "OpDh-S7u_XS8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "pd_mapping=pd_mapping.fillna('None')"
      ],
      "metadata": {
        "id": "_zv5uplz9ta6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "mapping_list=[]\n",
        "rap_lit=[]\n",
        "msu_list=[]\n",
        "for row in pd_mapping.to_numpy():\n",
        "  if row[1] != 'None' and row[1] != '_':\n",
        "    name = row[1].upper()\n",
        "    mapping_list.append([row[0],name])\n",
        "  elif row[2] != 'None' and row[2] != '_':\n",
        "    if ',' in row[2]:\n",
        "      name = row[2].split(',')[0]\n",
        "      name = name.upper()\n",
        "      mapping_list.append([row[0],name])\n",
        "    else:\n",
        "      name = row[2].upper()\n",
        "      mapping_list.append([row[0],name])\n",
        "\n",
        "df = pd.DataFrame(mapping_list, columns =['Gene_ID', 'Name'])"
      ],
      "metadata": {
        "id": "14eBuOOK90f3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df['Gene_ID'] = 'http://identifiers.org/oryzabase.gene/' + df['Gene_ID'].astype(str)"
      ],
      "metadata": {
        "id": "rZtjjRkIBNkN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df.head()"
      ],
      "metadata": {
        "id": "vDb0qfGw-g4z"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from pandas._libs import index\n",
        "df.to_csv('pubAnnotation.txt', sep='\\t',columns=['Name','Gene_ID'],index=False)"
      ],
      "metadata": {
        "id": "4PRxl-f_JIIl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Retrieving Title and Abstracts from PubMed"
      ],
      "metadata": {
        "id": "YS9bfR8D76Ju"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import Entrez\n",
        "import time\n",
        "Entrez.email = 'pierre.larmande@ird.fr'\n",
        "#pmids = [29090559 29058482 28991880 28984387 28862677 28804631 28801717 28770950 28768831 28707064 28701466 28685492 28623948 28551248]\n",
        "handle = Entrez.efetch(db=\"pubmed\", id=pmids[1:10], retmode=\"xml\")\n",
        "records = Entrez.read(handle)\n",
        "#abstracts = [pubmed_article['MedlineCitation']['Article']['Abstract']['AbstractText'][0]  if 'Abstract' in pubmed_article['MedlineCitation']['Article'].keys() else pubmed_article['MedlineCitation']['Article']['ArticleTitle']  for pubmed_article in records['PubmedArticle']]\n",
        "#abstract_dict = dict(zip(pmids, abstracts))\n",
        "#print abstract_dict\n"
      ],
      "metadata": {
        "id": "C6rYjjq4SzaI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "output_writer = open(abstracts_file, \"w\")"
      ],
      "metadata": {
        "id": "OdfPuJ98w-c9"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#abstracts = [pubmed_article['MedlineCitation']['Article']['Abstract']['AbstractText'][0]  if 'Abstract' in pubmed_article['MedlineCitation']['Article'].keys() else pubmed_article['MedlineCitation']['Article']['ArticleTitle']  for pubmed_article in records['PubmedArticle']]\n",
        "#abstract_dict = dict(zip(pmids, abstracts))\n",
        "#print(abstract_dict)\n",
        "for pubmed_article in records['PubmedArticle']:\n",
        "  pmid = pubmed_article['MedlineCitation']['PMID'][:]\n",
        "  title = pubmed_article['MedlineCitation']['Article']['ArticleTitle']\n",
        "  abstract = pubmed_article['MedlineCitation']['Article']['Abstract']['AbstractText'][0]  if 'Abstract' in pubmed_article['MedlineCitation']['Article'].keys() else ''\n",
        "  #print(pmid)\n",
        "  output_writer.write(pmid + '\\t' + title + '\\t' + abstract +'\\n')\n",
        "\n",
        "handle.close()\n",
        "output_writer.close"
      ],
      "metadata": {
        "id": "Cni7JTkywopv",
        "outputId": "f47521e3-f2d0-4a4f-f290-7d76cc827682",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<function TextIOWrapper.close>"
            ]
          },
          "metadata": {},
          "execution_count": 52
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "scbTFDg4rviy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# NER Tagging"
      ],
      "metadata": {
        "id": "4NEG_vrz416q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from flair.data import Sentence\n",
        "from flair.models import MultiTagger\n",
        "from flair.tokenization import SciSpacyTokenizer\n",
        "from flair.tokenization import SciSpacySentenceSplitter\n",
        "import en_core_sci_sm\n",
        "nlp = en_core_sci_sm.load()"
      ],
      "metadata": {
        "id": "QOqlZ6_CrxPs"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# load biomedical tagger\n",
        "tagger = MultiTagger.load(\"hunflair\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sSeKKFcst9Fr",
        "outputId": "703963d2-027c-459a-99bb-dc9b228ff486"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:25:56,364 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_smallish_models/cellline/hunflair-celline-v1.0.pt not found in cache, downloading to /tmp/tmp27ao_ls7\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104886372/1104886372 [01:01<00:00, 17965486.50B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:26:58,357 copying /tmp/tmp27ao_ls7 to cache at /root/.flair/models/hunflair-celline-v1.0.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:27:02,301 removing temp file /tmp/tmp27ao_ls7\n",
            "2022-09-01 08:27:02,448 loading file /root/.flair/models/hunflair-celline-v1.0.pt\n",
            "2022-09-01 08:27:10,022 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-CellLine, B-CellLine, I-CellLine, E-CellLine, <START>, <STOP>\n",
            "2022-09-01 08:27:11,272 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_allcorpus_models/huner-chemical/hunflair-chemical-full-v1.0.pt not found in cache, downloading to /tmp/tmp3rqcr0dp\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104888221/1104888221 [01:03<00:00, 17282326.57B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:28:15,699 copying /tmp/tmp3rqcr0dp to cache at /root/.flair/models/hunflair-chemical-full-v1.0.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:28:20,465 removing temp file /tmp/tmp3rqcr0dp\n",
            "2022-09-01 08:28:20,632 loading file /root/.flair/models/hunflair-chemical-full-v1.0.pt\n",
            "2022-09-01 08:28:28,367 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Chemical, B-Chemical, I-Chemical, E-Chemical, <START>, <STOP>\n",
            "2022-09-01 08:28:29,455 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_allcorpus_models/huner-disease/hunflair-disease-full-v1.0.pt not found in cache, downloading to /tmp/tmps1w_35iy\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104886193/1104886193 [01:01<00:00, 17933870.56B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:29:31,555 copying /tmp/tmps1w_35iy to cache at /root/.flair/models/hunflair-disease-full-v1.0.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:29:35,689 removing temp file /tmp/tmps1w_35iy\n",
            "2022-09-01 08:29:35,832 loading file /root/.flair/models/hunflair-disease-full-v1.0.pt\n",
            "2022-09-01 08:29:43,004 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, B-Disease, E-Disease, I-Disease, S-Disease, <START>, <STOP>\n",
            "2022-09-01 08:29:44,251 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_allcorpus_models/huner-gene/hunflair-gene-full-v1.0.pt not found in cache, downloading to /tmp/tmpgtg5060x\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104887211/1104887211 [01:01<00:00, 18028250.88B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:30:46,041 copying /tmp/tmpgtg5060x to cache at /root/.flair/models/hunflair-gene-full-v1.0.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:30:49,896 removing temp file /tmp/tmpgtg5060x\n",
            "2022-09-01 08:30:50,075 loading file /root/.flair/models/hunflair-gene-full-v1.0.pt\n",
            "2022-09-01 08:30:58,177 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Gene, B-Gene, I-Gene, E-Gene, <START>, <STOP>\n",
            "2022-09-01 08:30:59,273 https://nlp.informatik.hu-berlin.de/resources/models/hunflair_allcorpus_models/huner-species/hunflair-species-full-v1.1.pt not found in cache, downloading to /tmp/tmprah139kd\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 1104886931/1104886931 [01:01<00:00, 17917590.00B/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:32:01,431 copying /tmp/tmprah139kd to cache at /root/.flair/models/hunflair-species-full-v1.1.pt\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "2022-09-01 08:32:05,374 removing temp file /tmp/tmprah139kd\n",
            "2022-09-01 08:32:05,538 loading file /root/.flair/models/hunflair-species-full-v1.1.pt\n",
            "2022-09-01 08:32:12,799 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Species, B-Species, I-Species, E-Species, <START>, <STOP>\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# initialize the sentence splitter\n",
        "splitter = SciSpacySentenceSplitter()"
      ],
      "metadata": {
        "id": "rU2q0SoqC80c"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "abstracts = pd.read_csv(abstracts_file, sep='\\t', delimiter=None, names=['PMID', 'Title', 'Abstract'], dtype='str', index_col=['PMID'], encoding = \"ISO-8859-1\")"
      ],
      "metadata": {
        "id": "VCrsrfGc4zvS"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "LAYYmv8MC79E"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "abstracts.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        },
        "id": "KvweA205rWiT",
        "outputId": "6de1c631-a160-41dc-dff5-d07ac4b1c8aa"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                      Title  \\\n",
              "PMID                                                          \n",
              "31424513  Phosphoproteomic Profiling Reveals the Importa...   \n",
              "12582914  Substitution mapping of Pup1: a major QTL incr...   \n",
              "22020099  Rice GDP dissociation inhibitor 3 inhibits OsM...   \n",
              "12582916  Genetic basis of hybrid breakdown in a Japonic...   \n",
              "25821195  Quantitative trait locus analysis and fine map...   \n",
              "\n",
              "                                                   Abstract  \n",
              "PMID                                                         \n",
              "31424513  Phosphorus is one of the most important macron...  \n",
              "12582914  A major QTL for P uptake had previously been m...  \n",
              "22020099  GDP dissociation inhibitor (GDI) plays an esse...  \n",
              "12582916  Reproductive barriers often arise in hybrid pr...  \n",
              "25821195  Two QTLs were identified to control panicle le...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-9aa88615-9506-49c9-b48e-d470a6ce9089\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Title</th>\n",
              "      <th>Abstract</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PMID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>31424513</th>\n",
              "      <td>Phosphoproteomic Profiling Reveals the Importa...</td>\n",
              "      <td>Phosphorus is one of the most important macron...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12582914</th>\n",
              "      <td>Substitution mapping of Pup1: a major QTL incr...</td>\n",
              "      <td>A major QTL for P uptake had previously been m...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22020099</th>\n",
              "      <td>Rice GDP dissociation inhibitor 3 inhibits OsM...</td>\n",
              "      <td>GDP dissociation inhibitor (GDI) plays an esse...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12582916</th>\n",
              "      <td>Genetic basis of hybrid breakdown in a Japonic...</td>\n",
              "      <td>Reproductive barriers often arise in hybrid pr...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25821195</th>\n",
              "      <td>Quantitative trait locus analysis and fine map...</td>\n",
              "      <td>Two QTLs were identified to control panicle le...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9aa88615-9506-49c9-b48e-d470a6ce9089')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-9aa88615-9506-49c9-b48e-d470a6ce9089 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-9aa88615-9506-49c9-b48e-d470a6ce9089');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "short_list = abstracts.head().copy()"
      ],
      "metadata": {
        "id": "r11RPj13qGcK"
      },
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "short_list.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 238
        },
        "id": "NDzUT2YJvrhR",
        "outputId": "ce63d014-4285-4c2a-ea90-214ed23dd1ad"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                                      Title  \\\n",
              "PMID                                                          \n",
              "31424513  Phosphoproteomic Profiling Reveals the Importa...   \n",
              "12582914  Substitution mapping of Pup1: a major QTL incr...   \n",
              "22020099  Rice GDP dissociation inhibitor 3 inhibits OsM...   \n",
              "12582916  Genetic basis of hybrid breakdown in a Japonic...   \n",
              "25821195  Quantitative trait locus analysis and fine map...   \n",
              "\n",
              "                                                   Abstract  \n",
              "PMID                                                         \n",
              "31424513  Phosphorus is one of the most important macron...  \n",
              "12582914  A major QTL for P uptake had previously been m...  \n",
              "22020099  GDP dissociation inhibitor (GDI) plays an esse...  \n",
              "12582916  Reproductive barriers often arise in hybrid pr...  \n",
              "25821195  Two QTLs were identified to control panicle le...  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3cfa0a65-dfd9-4c7d-bbff-83059c9a899c\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Title</th>\n",
              "      <th>Abstract</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PMID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>31424513</th>\n",
              "      <td>Phosphoproteomic Profiling Reveals the Importa...</td>\n",
              "      <td>Phosphorus is one of the most important macron...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12582914</th>\n",
              "      <td>Substitution mapping of Pup1: a major QTL incr...</td>\n",
              "      <td>A major QTL for P uptake had previously been m...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22020099</th>\n",
              "      <td>Rice GDP dissociation inhibitor 3 inhibits OsM...</td>\n",
              "      <td>GDP dissociation inhibitor (GDI) plays an esse...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12582916</th>\n",
              "      <td>Genetic basis of hybrid breakdown in a Japonic...</td>\n",
              "      <td>Reproductive barriers often arise in hybrid pr...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25821195</th>\n",
              "      <td>Quantitative trait locus analysis and fine map...</td>\n",
              "      <td>Two QTLs were identified to control panicle le...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3cfa0a65-dfd9-4c7d-bbff-83059c9a899c')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-3cfa0a65-dfd9-4c7d-bbff-83059c9a899c button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-3cfa0a65-dfd9-4c7d-bbff-83059c9a899c');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "output_writer = open(\"output/annotated_abstracts.tsv\", \"w\")"
      ],
      "metadata": {
        "id": "Qv9c1PgjVBUe"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#import en_core_sci_sm\n",
        "#nlp = en_core_sci_sm.load()\n",
        "for index, record in short_list.iterrows():\n",
        "  sentence = Sentence(record['Title'], use_tokenizer=SciSpacyTokenizer())\n",
        "  tagger.predict(sentence)\n",
        "  print(\"\\nPubMed: {}\\n\\nTitle: {}\\n\".format(index,record['Title']))\n",
        "  #print(sentence.to_dict())\n",
        "  for annotation_layer in sentence.annotation_layers.keys():\n",
        "    for entity in sentence.get_spans(annotation_layer):\n",
        "      print(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\".format(index,record['Title'],entity.text,entity.labels,entity.tag,entity.start_position,entity.end_position,entity.score))\n",
        "      output_writer.write(\"{}\\t{}\\t{}\\n\".format(index,record['Title'],entity.labels))\n",
        "  # split text into a list of Sentence objects\n",
        "  sentences = splitter.split(record['Abstract'])\n",
        "  # you can apply the HunFlair tagger directly to this list\n",
        "  tagger.predict(sentences)\n",
        "  for sentence in sentences:\n",
        "    print(\"\\n\")\n",
        "    #print(sentence.to_tagged_string())\n",
        "    for annotation_layer in sentence.annotation_layers.keys():\n",
        "      for entity in sentence.get_spans(annotation_layer):\n",
        "        print(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\".format(index,sentence.text,entity.text,entity.labels,entity.tag,entity.start_position,entity.end_position,entity.score))\n",
        "        output_writer.write(\"{}\\t{}\\t{}\\n\".format(index,sentence.text,entity.labels))\n",
        "\n",
        "output_writer.close()\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "OT3ir3lCrtm9",
        "outputId": "2047408b-d626-432f-a20d-50f7cacb7da5"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "PubMed: 31424513\n",
            "\n",
            "Title: Phosphoproteomic Profiling Reveals the Importance of CK2, MAPKs and CDPKs in Response to Phosphate Starvation in Rice.\n",
            "\n",
            "31424513\tPhosphoproteomic Profiling Reveals the Importance of CK2, MAPKs and CDPKs in Response to Phosphate Starvation in Rice.\tPhosphate\t['Span[14:15]: \"Phosphate\"'/'Chemical' (0.9474)]\tChemical\t89\t98\t0.9473901987075806\n",
            "31424513\tPhosphoproteomic Profiling Reveals the Importance of CK2, MAPKs and CDPKs in Response to Phosphate Starvation in Rice.\tCK2\t['Span[6:7]: \"CK2\"'/'Gene' (0.9083)]\tGene\t53\t56\t0.9083380103111267\n",
            "31424513\tPhosphoproteomic Profiling Reveals the Importance of CK2, MAPKs and CDPKs in Response to Phosphate Starvation in Rice.\tMAPKs\t['Span[8:9]: \"MAPKs\"'/'Gene' (0.8176)]\tGene\t58\t63\t0.8175804615020752\n",
            "31424513\tPhosphoproteomic Profiling Reveals the Importance of CK2, MAPKs and CDPKs in Response to Phosphate Starvation in Rice.\tCDPKs\t['Span[10:11]: \"CDPKs\"'/'Gene' (0.9272)]\tGene\t68\t73\t0.9272241592407227\n",
            "31424513\tPhosphoproteomic Profiling Reveals the Importance of CK2, MAPKs and CDPKs in Response to Phosphate Starvation in Rice.\tRice\t['Span[17:18]: \"Rice\"'/'Species' (0.832)]\tSpecies\t113\t117\t0.8320468068122864\n",
            "\n",
            "\n",
            "31424513\tPhosphorus is one of the most important macronutrients required for plant growth and development .\tPhosphorus\t['Span[0:1]: \"Phosphorus\"'/'Chemical' (0.9813)]\tChemical\t0\t10\t0.981296956539154\n",
            "\n",
            "\n",
            "31424513\tThe importance of phosphorylation modification in regulating phosphate ( Pi ) homeostasis in plants is emerging .\tphosphate\t['Span[7:8]: \"phosphate\"'/'Chemical' (0.9507)]\tChemical\t61\t70\t0.9507336020469666\n",
            "\n",
            "\n",
            "31424513\tWe performed phosphoproteomic profiling to characterize proteins whose degree of phosphorylation is altered in response to Pi starvation in rice root .\trice\t['Span[19:20]: \"rice\"'/'Species' (0.8734)]\tSpecies\t140\t144\t0.8734029531478882\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "31424513\tKyoto Encyclopedia of Genes and Genomes ( KEGG ) pathway analysis with the differentially phosphorylated proteins indicated that RNA processing , transport , splicing and translation and carbon metabolism played critical roles in response to Pi starvation in rice .\tcarbon\t['Span[27:28]: \"carbon\"'/'Chemical' (0.9602)]\tChemical\t183\t189\t0.9601708054542542\n",
            "31424513\tKyoto Encyclopedia of Genes and Genomes ( KEGG ) pathway analysis with the differentially phosphorylated proteins indicated that RNA processing , transport , splicing and translation and carbon metabolism played critical roles in response to Pi starvation in rice .\trice\t['Span[38:39]: \"rice\"'/'Species' (0.8153)]\tSpecies\t255\t259\t0.8152893781661987\n",
            "\n",
            "\n",
            "31424513\tLevels of phosphorylation of four mitogen - activated protein kinases ( MAPKs ) , including OsMAPK6 , five calcium - dependent protein kinases ( CDPKs ) and OsCK2Î²3 decreased in response to Pi starvation .\tcalcium\t['Span[18:19]: \"calcium\"'/'Chemical' (0.7333)]\tChemical\t101\t108\t0.7332855463027954\n",
            "31424513\tLevels of phosphorylation of four mitogen - activated protein kinases ( MAPKs ) , including OsMAPK6 , five calcium - dependent protein kinases ( CDPKs ) and OsCK2Î²3 decreased in response to Pi starvation .\tmitogen - activated protein kinases\t['Span[5:10]: \"mitogen - activated protein kinases\"'/'Gene' (0.6933)]\tGene\t34\t67\t0.6932746648788453\n",
            "31424513\tLevels of phosphorylation of four mitogen - activated protein kinases ( MAPKs ) , including OsMAPK6 , five calcium - dependent protein kinases ( CDPKs ) and OsCK2Î²3 decreased in response to Pi starvation .\tMAPKs\t['Span[11:12]: \"MAPKs\"'/'Gene' (0.9207)]\tGene\t69\t74\t0.9207105040550232\n",
            "31424513\tLevels of phosphorylation of four mitogen - activated protein kinases ( MAPKs ) , including OsMAPK6 , five calcium - dependent protein kinases ( CDPKs ) and OsCK2Î²3 decreased in response to Pi starvation .\tOsMAPK6\t['Span[15:16]: \"OsMAPK6\"'/'Gene' (0.9508)]\tGene\t87\t94\t0.95076984167099\n",
            "31424513\tLevels of phosphorylation of four mitogen - activated protein kinases ( MAPKs ) , including OsMAPK6 , five calcium - dependent protein kinases ( CDPKs ) and OsCK2Î²3 decreased in response to Pi starvation .\tcalcium - dependent protein kinases\t['Span[18:23]: \"calcium - dependent protein kinases\"'/'Gene' (0.7019)]\tGene\t101\t134\t0.7019137024879456\n",
            "31424513\tLevels of phosphorylation of four mitogen - activated protein kinases ( MAPKs ) , including OsMAPK6 , five calcium - dependent protein kinases ( CDPKs ) and OsCK2Î²3 decreased in response to Pi starvation .\tCDPKs\t['Span[24:25]: \"CDPKs\"'/'Gene' (0.9065)]\tGene\t136\t141\t0.9065105319023132\n",
            "31424513\tLevels of phosphorylation of four mitogen - activated protein kinases ( MAPKs ) , including OsMAPK6 , five calcium - dependent protein kinases ( CDPKs ) and OsCK2Î²3 decreased in response to Pi starvation .\tOsCK2Î²3\t['Span[27:28]: \"OsCK2Î²3\"'/'Gene' (0.9603)]\tGene\t147\t155\t0.9602931141853333\n",
            "\n",
            "\n",
            "31424513\tThe decreased phosphorylation level of OsMAPK6 was confirmed by Western blotting .\tOsMAPK6\t['Span[5:6]: \"OsMAPK6\"'/'Gene' (0.9637)]\tGene\t39\t46\t0.963679313659668\n",
            "\n",
            "\n",
            "31424513\tMutation of OsMAPK6 led to Pi accumulation under Pi - sufficient conditions .\tOsMAPK6\t['Span[2:3]: \"OsMAPK6\"'/'Gene' (0.9642)]\tGene\t12\t19\t0.964155375957489\n",
            "\n",
            "\n",
            "31424513\tMotif analysis indicated that the putative MAPK , casein kinase 2 ( CK2 ) and CDPK substrates represented about 54.4 % , 21.5 % and 4.7 % , respectively , of the proteins exhibiting differential phosphorylation .\tMAPK\t['Span[6:7]: \"MAPK\"'/'Gene' (0.859)]\tGene\t43\t47\t0.8590115904808044\n",
            "31424513\tMotif analysis indicated that the putative MAPK , casein kinase 2 ( CK2 ) and CDPK substrates represented about 54.4 % , 21.5 % and 4.7 % , respectively , of the proteins exhibiting differential phosphorylation .\tcasein kinase 2\t['Span[8:11]: \"casein kinase 2\"'/'Gene' (0.7095)]\tGene\t49\t64\t0.7095002134641012\n",
            "31424513\tMotif analysis indicated that the putative MAPK , casein kinase 2 ( CK2 ) and CDPK substrates represented about 54.4 % , 21.5 % and 4.7 % , respectively , of the proteins exhibiting differential phosphorylation .\tCK2\t['Span[12:13]: \"CK2\"'/'Gene' (0.9663)]\tGene\t66\t69\t0.9662994742393494\n",
            "31424513\tMotif analysis indicated that the putative MAPK , casein kinase 2 ( CK2 ) and CDPK substrates represented about 54.4 % , 21.5 % and 4.7 % , respectively , of the proteins exhibiting differential phosphorylation .\tCDPK\t['Span[15:16]: \"CDPK\"'/'Gene' (0.8566)]\tGene\t75\t79\t0.8566447496414185\n",
            "\n",
            "\n",
            "31424513\tBased on the motif analysis , 191 , 151 and 46 candidate substrates for MAPK , CK2 and CDPK were identified .\tMAPK\t['Span[14:15]: \"MAPK\"'/'Gene' (0.9263)]\tGene\t70\t74\t0.9262799620628357\n",
            "31424513\tBased on the motif analysis , 191 , 151 and 46 candidate substrates for MAPK , CK2 and CDPK were identified .\tCK2\t['Span[16:17]: \"CK2\"'/'Gene' (0.9519)]\tGene\t76\t79\t0.9519212245941162\n",
            "31424513\tBased on the motif analysis , 191 , 151 and 46 candidate substrates for MAPK , CK2 and CDPK were identified .\tCDPK\t['Span[18:19]: \"CDPK\"'/'Gene' (0.9616)]\tGene\t84\t88\t0.9616343975067139\n",
            "\n",
            "\n",
            "31424513\tThese results indicate that modification of phosphorylation profiles provides complementary information on Pi - starvation - induced processes , with CK2 , MAPK and CDPK protein kinase families playing key roles in these processes in rice .\tCK2\t['Span[20:21]: \"CK2\"'/'Gene' (0.8855)]\tGene\t145\t148\t0.8854954242706299\n",
            "31424513\tThese results indicate that modification of phosphorylation profiles provides complementary information on Pi - starvation - induced processes , with CK2 , MAPK and CDPK protein kinase families playing key roles in these processes in rice .\tMAPK\t['Span[22:23]: \"MAPK\"'/'Gene' (0.8507)]\tGene\t150\t154\t0.8506781458854675\n",
            "31424513\tThese results indicate that modification of phosphorylation profiles provides complementary information on Pi - starvation - induced processes , with CK2 , MAPK and CDPK protein kinase families playing key roles in these processes in rice .\tCDPK protein kinase families\t['Span[24:28]: \"CDPK protein kinase families\"'/'Gene' (0.622)]\tGene\t159\t187\t0.6220401376485825\n",
            "31424513\tThese results indicate that modification of phosphorylation profiles provides complementary information on Pi - starvation - induced processes , with CK2 , MAPK and CDPK protein kinase families playing key roles in these processes in rice .\trice\t['Span[35:36]: \"rice\"'/'Species' (0.8182)]\tSpecies\t228\t232\t0.8181636929512024\n",
            "\n",
            "PubMed: 12582914\n",
            "\n",
            "Title: Substitution mapping of Pup1: a major QTL increasing phosphorus uptake of rice from a phosphorus-deficient soil.\n",
            "\n",
            "12582914\tSubstitution mapping of Pup1: a major QTL increasing phosphorus uptake of rice from a phosphorus-deficient soil.\tphosphorus\t['Span[9:10]: \"phosphorus\"'/'Chemical' (0.9763)]\tChemical\t53\t63\t0.9763320684432983\n",
            "12582914\tSubstitution mapping of Pup1: a major QTL increasing phosphorus uptake of rice from a phosphorus-deficient soil.\tphosphorus\t['Span[15:16]: \"phosphorus\"'/'Chemical' (0.9221)]\tChemical\t86\t96\t0.9221430420875549\n",
            "12582914\tSubstitution mapping of Pup1: a major QTL increasing phosphorus uptake of rice from a phosphorus-deficient soil.\tPup1\t['Span[3:4]: \"Pup1\"'/'Gene' (0.9566)]\tGene\t24\t28\t0.9566439390182495\n",
            "12582914\tSubstitution mapping of Pup1: a major QTL increasing phosphorus uptake of rice from a phosphorus-deficient soil.\trice\t['Span[12:13]: \"rice\"'/'Species' (0.8948)]\tSpecies\t74\t78\t0.8948140740394592\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "12582914\tQTL mapping showed that close to 80 % of the variation between families was due to a single QTL , hereafter referred to as Pup1 ( Phosphorus uptake 1 ) .\tPhosphorus\t['Span[26:27]: \"Phosphorus\"'/'Chemical' (0.8537)]\tChemical\t127\t137\t0.853735625743866\n",
            "12582914\tQTL mapping showed that close to 80 % of the variation between families was due to a single QTL , hereafter referred to as Pup1 ( Phosphorus uptake 1 ) .\tPup1\t['Span[24:25]: \"Pup1\"'/'Gene' (0.8106)]\tGene\t121\t125\t0.8106302618980408\n",
            "\n",
            "\n",
            "12582914\tPup1 was placed in a 3-cM interval flanked by markers S14025 and S13126 , which is within 1 cM of the position identified in the original QTL mapping experiment .\tPup1\t['Span[0:1]: \"Pup1\"'/'Gene' (0.8562)]\tGene\t0\t4\t0.8562453985214233\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "12582914\tSubstitution mapping revealed that Pup1 co - segregated with marker S13126 and that the flanking markers , S14025 and S13752 , were outside the interval containing Pup1 .\tPup1\t['Span[4:5]: \"Pup1\"'/'Gene' (0.8521)]\tGene\t35\t39\t0.8521330952644348\n",
            "12582914\tSubstitution mapping revealed that Pup1 co - segregated with marker S13126 and that the flanking markers , S14025 and S13752 , were outside the interval containing Pup1 .\tPup1\t['Span[26:27]: \"Pup1\"'/'Gene' (0.873)]\tGene\t160\t164\t0.8730087876319885\n",
            "\n",
            "\n",
            "12582914\tThe two mapping strategies therefore yielded almost identical results and , in combining the advantages of both , Pup1 could be mapped with high certainty .\tPup1\t['Span[18:19]: \"Pup1\"'/'Gene' (0.8392)]\tGene\t112\t116\t0.8392370343208313\n",
            "\n",
            "\n",
            "\n",
            "PubMed: 22020099\n",
            "\n",
            "Title: Rice GDP dissociation inhibitor 3 inhibits OsMAPK2 activity through physical interaction.\n",
            "\n",
            "22020099\tRice GDP dissociation inhibitor 3 inhibits OsMAPK2 activity through physical interaction.\tGDP\t['Span[1:2]: \"GDP\"'/'Chemical' (0.8684)]\tChemical\t5\t8\t0.8684031963348389\n",
            "22020099\tRice GDP dissociation inhibitor 3 inhibits OsMAPK2 activity through physical interaction.\tRice GDP dissociation inhibitor 3\t['Span[0:5]: \"Rice GDP dissociation inhibitor 3\"'/'Gene' (0.687)]\tGene\t0\t33\t0.6869501590728759\n",
            "22020099\tRice GDP dissociation inhibitor 3 inhibits OsMAPK2 activity through physical interaction.\tOsMAPK2\t['Span[6:7]: \"OsMAPK2\"'/'Gene' (0.9456)]\tGene\t43\t50\t0.9456453323364258\n",
            "\n",
            "\n",
            "22020099\tGDP dissociation inhibitor ( GDI ) plays an essential role in regulating the state of bound nucleotides and subcellular localizations of Rab proteins .\tGDP\t['Span[0:1]: \"GDP\"'/'Chemical' (0.9571)]\tChemical\t0\t3\t0.9571399688720703\n",
            "22020099\tGDP dissociation inhibitor ( GDI ) plays an essential role in regulating the state of bound nucleotides and subcellular localizations of Rab proteins .\tnucleotides\t['Span[16:17]: \"nucleotides\"'/'Chemical' (0.866)]\tChemical\t90\t101\t0.8659579157829285\n",
            "22020099\tGDP dissociation inhibitor ( GDI ) plays an essential role in regulating the state of bound nucleotides and subcellular localizations of Rab proteins .\tGDP dissociation inhibitor\t['Span[0:3]: \"GDP dissociation inhibitor\"'/'Gene' (0.7329)]\tGene\t0\t26\t0.7329129775365194\n",
            "22020099\tGDP dissociation inhibitor ( GDI ) plays an essential role in regulating the state of bound nucleotides and subcellular localizations of Rab proteins .\tGDI\t['Span[4:5]: \"GDI\"'/'Gene' (0.8254)]\tGene\t28\t31\t0.8253921270370483\n",
            "22020099\tGDP dissociation inhibitor ( GDI ) plays an essential role in regulating the state of bound nucleotides and subcellular localizations of Rab proteins .\tRab\t['Span[21:22]: \"Rab\"'/'Gene' (0.7816)]\tGene\t135\t138\t0.7816373109817505\n",
            "\n",
            "\n",
            "22020099\tIn our previous study , we showed that OsGDI3 facilitates the recycling of OsRab11 with a help of OsGAP1 .\tOsGDI3\t['Span[8:9]: \"OsGDI3\"'/'Gene' (0.825)]\tGene\t38\t44\t0.8249643445014954\n",
            "22020099\tIn our previous study , we showed that OsGDI3 facilitates the recycling of OsRab11 with a help of OsGAP1 .\tOsRab11\t['Span[13:14]: \"OsRab11\"'/'Gene' (0.9035)]\tGene\t74\t81\t0.9035075902938843\n",
            "22020099\tIn our previous study , we showed that OsGDI3 facilitates the recycling of OsRab11 with a help of OsGAP1 .\tOsGAP1\t['Span[18:19]: \"OsGAP1\"'/'Gene' (0.924)]\tGene\t97\t103\t0.924015998840332\n",
            "\n",
            "\n",
            "22020099\tIn this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\tOsGDI3\t['Span[7:8]: \"OsGDI3\"'/'Gene' (0.9153)]\tGene\t28\t34\t0.9152601361274719\n",
            "22020099\tIn this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\tsec19 - 1\t['Span[11:14]: \"sec19 - 1\"'/'Gene' (0.6907)]\tGene\t56\t63\t0.6907098889350891\n",
            "22020099\tIn this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\tyeast GDI gene\t['Span[23:26]: \"yeast GDI gene\"'/'Gene' (0.5873)]\tGene\t110\t124\t0.5872795879840851\n",
            "22020099\tIn this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\tOsGDI3\t['Span[29:30]: \"OsGDI3\"'/'Gene' (0.8895)]\tGene\t142\t148\t0.8894975185394287\n",
            "22020099\tIn this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\tyeast GDI\t['Span[35:37]: \"yeast GDI\"'/'Gene' (0.6493)]\tGene\t177\t186\t0.6493258476257324\n",
            "22020099\tIn this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\tyeast\t['Span[10:11]: \"yeast\"'/'Species' (0.5229)]\tSpecies\t50\t55\t0.522919774055481\n",
            "22020099\tIn this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\tyeast\t['Span[23:24]: \"yeast\"'/'Species' (0.8049)]\tSpecies\t110\t115\t0.8048754930496216\n",
            "22020099\tIn this study , we show that OsGDI3 complement the yeast sec19 - 1 mutant , a temperature - sensitive allele of the yeast GDI gene , suggesting that OsGDI3 is a functional ortholog of yeast GDI .\tyeast\t['Span[35:36]: \"yeast\"'/'Species' (0.6893)]\tSpecies\t177\t182\t0.6892857551574707\n",
            "\n",
            "\n",
            "22020099\tTo obtain further knowledge on the function of OsGDI3 , candidate OsGDI3-interacting proteins were identified by yeast two - hybrid screens .\tOsGDI3\t['Span[8:9]: \"OsGDI3\"'/'Gene' (0.9118)]\tGene\t47\t53\t0.9117977619171143\n",
            "22020099\tTo obtain further knowledge on the function of OsGDI3 , candidate OsGDI3-interacting proteins were identified by yeast two - hybrid screens .\tOsGDI3-interacting\t['Span[11:12]: \"OsGDI3-interacting\"'/'Gene' (0.7388)]\tGene\t65\t83\t0.7387580275535583\n",
            "\n",
            "\n",
            "22020099\tOsMAPK2 is one of OsGDI3 interacting proteins from yeast two - hybrid screens and subject to further analysis .\tOsMAPK2\t['Span[0:1]: \"OsMAPK2\"'/'Gene' (0.9493)]\tGene\t0\t7\t0.9493345022201538\n",
            "22020099\tOsMAPK2 is one of OsGDI3 interacting proteins from yeast two - hybrid screens and subject to further analysis .\tOsGDI3\t['Span[4:5]: \"OsGDI3\"'/'Gene' (0.8251)]\tGene\t18\t24\t0.8250884413719177\n",
            "\n",
            "\n",
            "22020099\tA kinase assay showed that the autophosphorylation activity of OsMAPK2 is inhibited by OsGDI3 in vitro .\tkinase\t['Span[1:2]: \"kinase\"'/'Gene' (0.6095)]\tGene\t2\t8\t0.6094748973846436\n",
            "22020099\tA kinase assay showed that the autophosphorylation activity of OsMAPK2 is inhibited by OsGDI3 in vitro .\tOsMAPK2\t['Span[9:10]: \"OsMAPK2\"'/'Gene' (0.9646)]\tGene\t63\t70\t0.9645830988883972\n",
            "22020099\tA kinase assay showed that the autophosphorylation activity of OsMAPK2 is inhibited by OsGDI3 in vitro .\tOsGDI3\t['Span[13:14]: \"OsGDI3\"'/'Gene' (0.7082)]\tGene\t87\t93\t0.7082422375679016\n",
            "\n",
            "\n",
            "22020099\tIn addition , ectopic expressions of OsGDI3-in Arabidopsis cause reductions at the level of phosphorylated AtMPK in phosphorylation activity .\tOsGDI3-in\t['Span[6:7]: \"OsGDI3-in\"'/'Gene' (0.7466)]\tGene\t36\t45\t0.7465742826461792\n",
            "22020099\tIn addition , ectopic expressions of OsGDI3-in Arabidopsis cause reductions at the level of phosphorylated AtMPK in phosphorylation activity .\tAtMPK\t['Span[15:16]: \"AtMPK\"'/'Gene' (0.8348)]\tGene\t106\t111\t0.8347622156143188\n",
            "22020099\tIn addition , ectopic expressions of OsGDI3-in Arabidopsis cause reductions at the level of phosphorylated AtMPK in phosphorylation activity .\tArabidopsis\t['Span[7:8]: \"Arabidopsis\"'/'Species' (0.9459)]\tSpecies\t46\t57\t0.9458708763122559\n",
            "\n",
            "\n",
            "22020099\tTaken together , OsGDI3 functions as a negative regulator of OsMAPK2 through modulating its kinase activity .\tOsGDI3\t['Span[3:4]: \"OsGDI3\"'/'Gene' (0.9232)]\tGene\t16\t22\t0.9232063889503479\n",
            "22020099\tTaken together , OsGDI3 functions as a negative regulator of OsMAPK2 through modulating its kinase activity .\tOsMAPK2\t['Span[10:11]: \"OsMAPK2\"'/'Gene' (0.9543)]\tGene\t60\t67\t0.9543498158454895\n",
            "\n",
            "PubMed: 12582916\n",
            "\n",
            "Title: Genetic basis of hybrid breakdown in a Japonica/Indica cross of rice, Oryza sativa L.\n",
            "\n",
            "12582916\tGenetic basis of hybrid breakdown in a Japonica/Indica cross of rice, Oryza sativa L.\trice\t['Span[12:13]: \"rice\"'/'Species' (0.87)]\tSpecies\t64\t68\t0.8700008988380432\n",
            "12582916\tGenetic basis of hybrid breakdown in a Japonica/Indica cross of rice, Oryza sativa L.\tOryza sativa\t['Span[14:16]: \"Oryza sativa\"'/'Species' (0.8237)]\tSpecies\t70\t82\t0.8237447142601013\n",
            "\n",
            "\n",
            "12582916\tReproductive barriers often arise in hybrid progeny between two varietal groups of Asian cultivated rice ( Oryza sativa L. ) , Japonica and Indica .\trice\t['Span[14:15]: \"rice\"'/'Species' (0.6544)]\tSpecies\t100\t104\t0.6544346213340759\n",
            "12582916\tReproductive barriers often arise in hybrid progeny between two varietal groups of Asian cultivated rice ( Oryza sativa L. ) , Japonica and Indica .\tOryza sativa\t['Span[16:18]: \"Oryza sativa\"'/'Species' (0.899)]\tSpecies\t107\t119\t0.8989758789539337\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "12582916\tIt was found that the hybrid breakdown is caused by a set of two nuclear genes , which were symbolized as hwe1 and hwe2 .\thwe1\t['Span[21:22]: \"hwe1\"'/'Gene' (0.943)]\tGene\t105\t109\t0.9430165886878967\n",
            "12582916\tIt was found that the hybrid breakdown is caused by a set of two nuclear genes , which were symbolized as hwe1 and hwe2 .\thwe2\t['Span[23:24]: \"hwe2\"'/'Gene' (0.9832)]\tGene\t114\t118\t0.9831914305686951\n",
            "\n",
            "\n",
            "12582916\tThe parental varieties , Asominori and IR24 , carry hwe1 ( + ) hwe1 ( + ) hwe2hwe2 and hwe1hwe1hwe2 ( + ) hwe2 ( + ) genotypes , respectively , whereas the progenies that showed a weakness performance carry the double recessive genotype ( hwe1hwe1hwe2hwe2 ) .\tweakness\t['Span[37:38]: \"weakness\"'/'Disease' (0.7534)]\tDisease\t164\t172\t0.7534271478652954\n",
            "12582916\tThe parental varieties , Asominori and IR24 , carry hwe1 ( + ) hwe1 ( + ) hwe2hwe2 and hwe1hwe1hwe2 ( + ) hwe2 ( + ) genotypes , respectively , whereas the progenies that showed a weakness performance carry the double recessive genotype ( hwe1hwe1hwe2hwe2 ) .\thwe1\t['Span[9:10]: \"hwe1\"'/'Gene' (0.8085)]\tGene\t50\t54\t0.808542013168335\n",
            "12582916\tThe parental varieties , Asominori and IR24 , carry hwe1 ( + ) hwe1 ( + ) hwe2hwe2 and hwe1hwe1hwe2 ( + ) hwe2 ( + ) genotypes , respectively , whereas the progenies that showed a weakness performance carry the double recessive genotype ( hwe1hwe1hwe2hwe2 ) .\thwe1\t['Span[13:14]: \"hwe1\"'/'Gene' (0.6694)]\tGene\t58\t62\t0.669355034828186\n",
            "12582916\tThe parental varieties , Asominori and IR24 , carry hwe1 ( + ) hwe1 ( + ) hwe2hwe2 and hwe1hwe1hwe2 ( + ) hwe2 ( + ) genotypes , respectively , whereas the progenies that showed a weakness performance carry the double recessive genotype ( hwe1hwe1hwe2hwe2 ) .\thwe2hwe2\t['Span[17:18]: \"hwe2hwe2\"'/'Gene' (0.6243)]\tGene\t66\t74\t0.6242925524711609\n",
            "12582916\tThe parental varieties , Asominori and IR24 , carry hwe1 ( + ) hwe1 ( + ) hwe2hwe2 and hwe1hwe1hwe2 ( + ) hwe2 ( + ) genotypes , respectively , whereas the progenies that showed a weakness performance carry the double recessive genotype ( hwe1hwe1hwe2hwe2 ) .\thwe1hwe1hwe2\t['Span[19:20]: \"hwe1hwe1hwe2\"'/'Gene' (0.8728)]\tGene\t79\t91\t0.8728388547897339\n",
            "12582916\tThe parental varieties , Asominori and IR24 , carry hwe1 ( + ) hwe1 ( + ) hwe2hwe2 and hwe1hwe1hwe2 ( + ) hwe2 ( + ) genotypes , respectively , whereas the progenies that showed a weakness performance carry the double recessive genotype ( hwe1hwe1hwe2hwe2 ) .\thwe2\t['Span[23:24]: \"hwe2\"'/'Gene' (0.6285)]\tGene\t95\t99\t0.6284576058387756\n",
            "12582916\tThe parental varieties , Asominori and IR24 , carry hwe1 ( + ) hwe1 ( + ) hwe2hwe2 and hwe1hwe1hwe2 ( + ) hwe2 ( + ) genotypes , respectively , whereas the progenies that showed a weakness performance carry the double recessive genotype ( hwe1hwe1hwe2hwe2 ) .\thwe1hwe1hwe2hwe2\t['Span[45:46]: \"hwe1hwe1hwe2hwe2\"'/'Gene' (0.9061)]\tGene\t223\t239\t0.9060546159744263\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "12582916\tBased on linkage analysis with RFLP markers , the hwe1 locus was located between RFLP markers R1869 and S1437 on chromosome 12 and the hwe2 locus was located between R3192 and C1211 on chromosome 1 .\thwe1\t['Span[9:10]: \"hwe1\"'/'Gene' (0.9237)]\tGene\t49\t53\t0.9236705899238586\n",
            "12582916\tBased on linkage analysis with RFLP markers , the hwe1 locus was located between RFLP markers R1869 and S1437 on chromosome 12 and the hwe2 locus was located between R3192 and C1211 on chromosome 1 .\thwe2\t['Span[24:25]: \"hwe2\"'/'Gene' (0.8569)]\tGene\t134\t138\t0.8569232821464539\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "PubMed: 25821195\n",
            "\n",
            "Title: Quantitative trait locus analysis and fine mapping of the qPL6 locus for panicle length in rice.\n",
            "\n",
            "25821195\tQuantitative trait locus analysis and fine mapping of the qPL6 locus for panicle length in rice.\tqPL6\t['Span[9:10]: \"qPL6\"'/'Gene' (0.7651)]\tGene\t58\t62\t0.7651040554046631\n",
            "25821195\tQuantitative trait locus analysis and fine mapping of the qPL6 locus for panicle length in rice.\trice\t['Span[15:16]: \"rice\"'/'Species' (0.8427)]\tSpecies\t91\t95\t0.842677891254425\n",
            "\n",
            "\n",
            "25821195\tTwo QTLs were identified to control panicle length in rice backcross lines , and one QTL qPL6 was finely mapped with potential in high yield breeding .\trice\t['Span[9:10]: \"rice\"'/'Species' (0.8444)]\tSpecies\t54\t58\t0.8444440960884094\n",
            "\n",
            "\n",
            "25821195\tPanicle length ( PL ) is the key determinant of panicle architecture in rice , and strongly affects yield components , such as grain number per panicle .\trice\t['Span[13:14]: \"rice\"'/'Species' (0.7066)]\tSpecies\t70\t74\t0.7066009640693665\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "25821195\tIn this study , we performed quantitative trait locus ( QTL ) analysis for PL in four backcross populations derived from the cross of Nipponbare ( japonica ) and WS3 ( indica ) , a new plant type ( NPT ) variety .\tNipponbare ( japonica\t['Span[24:27]: \"Nipponbare ( japonica\"'/'Species' (0.5884)]\tSpecies\t131\t151\t0.5883543093999227\n",
            "25821195\tIn this study , we performed quantitative trait locus ( QTL ) analysis for PL in four backcross populations derived from the cross of Nipponbare ( japonica ) and WS3 ( indica ) , a new plant type ( NPT ) variety .\tWS3 ( indica\t['Span[29:32]: \"WS3 ( indica\"'/'Species' (0.6692)]\tSpecies\t157\t168\t0.6691728631655375\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "25821195\tNear - isogenic lines ( NILs ) were developed to evaluate their contribution to important agronomic traits .\tNILs\t['Span[5:6]: \"NILs\"'/'CellLine' (0.9627)]\tCellLine\t21\t25\t0.9626673460006714\n",
            "\n",
            "\n",
            "25821195\tWe found that qPL6 and qPL8 had additive effects on PL trait .\tqPL6\t['Span[3:4]: \"qPL6\"'/'Gene' (0.659)]\tGene\t14\t18\t0.6589837074279785\n",
            "25821195\tWe found that qPL6 and qPL8 had additive effects on PL trait .\tqPL8\t['Span[5:6]: \"qPL8\"'/'Gene' (0.8077)]\tGene\t23\t27\t0.80767822265625\n",
            "\n",
            "\n",
            "25821195\tFor the qPL6 locus , the WS3 allele also increased panicle primary and secondary branches and grain number per panicle .\tqPL6\t['Span[2:3]: \"qPL6\"'/'Gene' (0.8681)]\tGene\t8\t12\t0.8681218028068542\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "25821195\tBy analyzing key recombinants in two steps , the qPL6 locus was finely mapped to a 25-kb interval , and 3 candidate genes were identified .\tqPL6 locus\t['Span[9:11]: \"qPL6 locus\"'/'Gene' (0.7371)]\tGene\t48\t58\t0.7371040284633636\n",
            "\n",
            "\n",
            "25821195\tAccording to the single nucleotide polymorphisms ( SNPs ) within candidate genes , 5 dCaps markers were designed and used to get haplotypes of 96 modern Chinese varieties , which proved that qPL6 locus is differentiated between indica and temperate japonica varieties .\tqPL6\t['Span[32:33]: \"qPL6\"'/'Gene' (0.7974)]\tGene\t187\t191\t0.7973887920379639\n",
            "\n",
            "\n",
            "25821195\tTaken together , the superior qPL6 allele can be applied in rice breeding programs for large sink size , particularly for japonica varieties that originally lack the allele .\tqPL6\t['Span[5:6]: \"qPL6\"'/'Gene' (0.5891)]\tGene\t29\t33\t0.5890799164772034\n",
            "25821195\tTaken together , the superior qPL6 allele can be applied in rice breeding programs for large sink size , particularly for japonica varieties that originally lack the allele .\trice\t['Span[11:12]: \"rice\"'/'Species' (0.9486)]\tSpecies\t59\t63\t0.9486351609230042\n",
            "25821195\tTaken together , the superior qPL6 allele can be applied in rice breeding programs for large sink size , particularly for japonica varieties that originally lack the allele .\tjaponica\t['Span[21:22]: \"japonica\"'/'Species' (0.7401)]\tSpecies\t120\t128\t0.7400726675987244\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "mV4Hvq10xfpC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#**Saving the results**"
      ],
      "metadata": {
        "id": "H6JTGnTm_f0H"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! git pull\n",
        "#! git add output/iric_oryzabase_tf_annotation.pkl\n",
        "! git commit -m \"saving\"\n",
        "! git push"
      ],
      "metadata": {
        "id": "V3BtDBUs_l6D",
        "outputId": "b4235cd9-4192-4e5a-8655-347fc95c0410",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "remote: Enumerating objects: 7, done.\u001b[K\n",
            "remote: Counting objects:  14% (1/7)\u001b[K\rremote: Counting objects:  28% (2/7)\u001b[K\rremote: Counting objects:  42% (3/7)\u001b[K\rremote: Counting objects:  57% (4/7)\u001b[K\rremote: Counting objects:  71% (5/7)\u001b[K\rremote: Counting objects:  85% (6/7)\u001b[K\rremote: Counting objects: 100% (7/7)\u001b[K\rremote: Counting objects: 100% (7/7), done.\u001b[K\n",
            "remote: Compressing objects:  25% (1/4)\u001b[K\rremote: Compressing objects:  50% (2/4)\u001b[K\rremote: Compressing objects:  75% (3/4)\u001b[K\rremote: Compressing objects: 100% (4/4)\u001b[K\rremote: Compressing objects: 100% (4/4), done.\u001b[K\n",
            "remote: Total 4 (delta 2), reused 0 (delta 0), pack-reused 0\u001b[K\n",
            "Unpacking objects: 100% (4/4), done.\n",
            "From https://github.com/pierrelarmande/OryzaGP\n",
            "   ff97898..996676b  main       -> origin/main\n",
            "hint: Waiting for your editor to close the file... error: unable to start editor 'editor'\n",
            "Not committing merge; use 'git commit' to complete the merge.\n",
            "[main 39228c5] saving\n",
            "fatal: could not read Password for 'https://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@github.com': No such device or address\n"
          ]
        }
      ]
    }
  ]
}